
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../">
      
      
        <link rel="next" href="../../bin/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.3">
    
    
      
        <title>gnomonicus_lib - Gnomonicus Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.d7758b05.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    <body dir="ltr">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#gnomonicus_lib" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Gnomonicus Documentation" class="md-header__button md-logo" aria-label="Gnomonicus Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Gnomonicus Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              gnomonicus_lib
            
          </span>
        </div>
      </div>
    </div>
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/oxfordmmm/gnomonicus" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    gnomonicus
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../readme/" class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../" class="md-tabs__link">
          
  
  Code Reference

        </a>
      </li>
    
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../bin/" class="md-tabs__link">
        
  
    
  
  CLI

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Gnomonicus Documentation" class="md-nav__button md-logo" aria-label="Gnomonicus Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Gnomonicus Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/oxfordmmm/gnomonicus" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    gnomonicus
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../readme/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Code Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Code Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gnomonicus
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    gnomonicus_lib
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    gnomonicus_lib
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib" class="md-nav__link">
    <span class="md-ellipsis">
      gnomonicus_lib
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.InvalidMutationException" class="md-nav__link">
    <span class="md-ellipsis">
      InvalidMutationException
    </span>
  </a>
  
    <nav class="md-nav" aria-label="InvalidMutationException">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gnomonicus_lib.InvalidMutationException.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.count_nucleotide_changes" class="md-nav__link">
    <span class="md-ellipsis">
      count_nucleotide_changes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.epistasis" class="md-nav__link">
    <span class="md-ellipsis">
      epistasis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.getGenes" class="md-nav__link">
    <span class="md-ellipsis">
      getGenes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.getMutations" class="md-nav__link">
    <span class="md-ellipsis">
      getMutations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.get_minority_population_type" class="md-nav__link">
    <span class="md-ellipsis">
      get_minority_population_type
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.parse_grumpy_evidence" class="md-nav__link">
    <span class="md-ellipsis">
      parse_grumpy_evidence
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.populateEffects" class="md-nav__link">
    <span class="md-ellipsis">
      populateEffects
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.populateMutations" class="md-nav__link">
    <span class="md-ellipsis">
      populateMutations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.populateVariants" class="md-nav__link">
    <span class="md-ellipsis">
      populateVariants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.saveJSON" class="md-nav__link">
    <span class="md-ellipsis">
      saveJSON
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.subset_multis" class="md-nav__link">
    <span class="md-ellipsis">
      subset_multis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.write_mutations_csv" class="md-nav__link">
    <span class="md-ellipsis">
      write_mutations_csv
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../bin/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CLI
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib" class="md-nav__link">
    <span class="md-ellipsis">
      gnomonicus_lib
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.InvalidMutationException" class="md-nav__link">
    <span class="md-ellipsis">
      InvalidMutationException
    </span>
  </a>
  
    <nav class="md-nav" aria-label="InvalidMutationException">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gnomonicus_lib.InvalidMutationException.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.count_nucleotide_changes" class="md-nav__link">
    <span class="md-ellipsis">
      count_nucleotide_changes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.epistasis" class="md-nav__link">
    <span class="md-ellipsis">
      epistasis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.getGenes" class="md-nav__link">
    <span class="md-ellipsis">
      getGenes
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.getMutations" class="md-nav__link">
    <span class="md-ellipsis">
      getMutations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.get_minority_population_type" class="md-nav__link">
    <span class="md-ellipsis">
      get_minority_population_type
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.parse_grumpy_evidence" class="md-nav__link">
    <span class="md-ellipsis">
      parse_grumpy_evidence
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.populateEffects" class="md-nav__link">
    <span class="md-ellipsis">
      populateEffects
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.populateMutations" class="md-nav__link">
    <span class="md-ellipsis">
      populateMutations
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.populateVariants" class="md-nav__link">
    <span class="md-ellipsis">
      populateVariants
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.saveJSON" class="md-nav__link">
    <span class="md-ellipsis">
      saveJSON
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.subset_multis" class="md-nav__link">
    <span class="md-ellipsis">
      subset_multis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gnomonicus_lib.write_mutations_csv" class="md-nav__link">
    <span class="md-ellipsis">
      write_mutations_csv
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>gnomonicus_lib</h1>

<div class="doc doc-object doc-module">



<a id="gnomonicus_lib"></a>
    <div class="doc doc-contents first">

        <p>gnomonicus.py is a library providing functions which pull together output VCF of the Lodestone TB pipeline
    with a reference genome and a resistance catalogue, and utilise grumpy and
    piezo to produce variants, mutations and an antibiogram.</p>
<p>Based on sp3predict</p>








  <div class="doc doc-children">








<div class="doc doc-object doc-class">



<h2 id="gnomonicus_lib.InvalidMutationException" class="doc doc-heading">
            <code>InvalidMutationException</code>


</h2>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code>Exception</code></p>


        <p>Custom exception raised when an invalid mutation is detected</p>






              <details class="quote">
                <summary>Source code in <code>gnomonicus/gnomonicus_lib.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">InvalidMutationException</span><span class="p">(</span><span class="ne">Exception</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Custom exception raised when an invalid mutation is detected&quot;&quot;&quot;</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">mutation</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Raise this exception</span>

<span class="sd">        Args:</span>
<span class="sd">            gene (str): Name of the gene</span>
<span class="sd">            mutation (str): The invalid mutation</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">message</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">@</span><span class="si">{</span><span class="n">mutation</span><span class="si">}</span><span class="s2"> is not a valid mutation!&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="gnomonicus_lib.InvalidMutationException.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">mutation</span><span class="p">)</span></code>

</h3>


    <div class="doc doc-contents ">

        <p>Raise this exception</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>gene</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Name of the gene</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mutation</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The invalid mutation</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gnomonicus/gnomonicus_lib.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">mutation</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Raise this exception</span>

<span class="sd">    Args:</span>
<span class="sd">        gene (str): Name of the gene</span>
<span class="sd">        mutation (str): The invalid mutation</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">message</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">gene</span><span class="si">}</span><span class="s2">@</span><span class="si">{</span><span class="n">mutation</span><span class="si">}</span><span class="s2"> is not a valid mutation!&quot;</span>
    <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">message</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>


<div class="doc doc-object doc-function">


<h2 id="gnomonicus_lib.count_nucleotide_changes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">count_nucleotide_changes</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Count number of changes between ref and alt</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>ref</code>
            </td>
            <td>
                  <code>str | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Ref nucleotides</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>alt</code>
            </td>
            <td>
                  <code>str | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Alt nucleotides</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>int | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>int | None: SNP distance if SNP, None if ref/alt were None</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gnomonicus/gnomonicus_lib.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">count_nucleotide_changes</span><span class="p">(</span><span class="n">ref</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span> <span class="n">alt</span><span class="p">:</span> <span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span> <span class="o">|</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Count number of changes between ref and alt</span>

<span class="sd">    Args:</span>
<span class="sd">        ref (str | None): Ref nucleotides</span>
<span class="sd">        alt (str | None): Alt nucleotides</span>

<span class="sd">    Returns:</span>
<span class="sd">        int | None: SNP distance if SNP, None if ref/alt were None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">ref</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="n">alt</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">return</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">r</span><span class="p">,</span> <span class="n">a</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">ref</span><span class="p">,</span> <span class="n">alt</span><span class="p">)</span> <span class="k">if</span> <span class="n">r</span> <span class="o">!=</span> <span class="n">a</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="gnomonicus_lib.epistasis" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">epistasis</span><span class="p">(</span><span class="n">mutations</span><span class="p">,</span> <span class="n">resistanceCatalogue</span><span class="p">,</span> <span class="n">phenotype</span><span class="p">,</span> <span class="n">effects</span><span class="p">,</span> <span class="n">effectsCounter</span><span class="p">,</span> <span class="n">vcfStem</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Add epistasis rules, overriding existing predictions as required</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>mutations</code>
            </td>
            <td>
                  <code>list[tuple[str | None, str]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Sample's mutations. Items are tuples of (gene, mutation)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>resistanceCatalogue</code>
            </td>
            <td>
                  <code><span title="piezo.ResistanceCatalogue">ResistanceCatalogue</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Resistance catalogue to use</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>phenotype</code>
            </td>
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary of phenotypes</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>effects</code>
            </td>
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Dictionary of effects</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>effectsCounter</code>
            </td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Position to add to effects dict</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>vcfStem</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Stem of the VCF file. Used as the <code>uniqueid</code></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>int</code></td>            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Incremented effects counter</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<details class="implicit-returns" open>
  <summary>Implicit returns</summary>
  <p>phenotypes: dict is updated in place
effects: dict is updated in place</p>
</details>
            <details class="quote">
              <summary>Source code in <code>gnomonicus/gnomonicus_lib.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span>
<span class="normal">858</span>
<span class="normal">859</span>
<span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">epistasis</span><span class="p">(</span>
    <span class="n">mutations</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span>
    <span class="n">resistanceCatalogue</span><span class="p">:</span> <span class="n">piezo</span><span class="o">.</span><span class="n">ResistanceCatalogue</span><span class="p">,</span>
    <span class="n">phenotype</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span>
    <span class="n">effects</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span>
    <span class="n">effectsCounter</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="n">vcfStem</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Add epistasis rules, overriding existing predictions as required</span>

<span class="sd">    Args:</span>
<span class="sd">        mutations (list[tuple[str | None, str]]): Sample&#39;s mutations. Items are tuples of (gene, mutation)</span>
<span class="sd">        resistanceCatalogue (piezo.ResistanceCatalogue): Resistance catalogue to use</span>
<span class="sd">        phenotype (dict): Dictionary of phenotypes</span>
<span class="sd">        effects (dict): Dictionary of effects</span>
<span class="sd">        effectsCounter (int): Position to add to effects dict</span>
<span class="sd">        vcfStem (str): Stem of the VCF file. Used as the `uniqueid`</span>

<span class="sd">    Returns:</span>
<span class="sd">        int: Incremented effects counter</span>

<span class="sd">    Implicit returns:</span>
<span class="sd">        phenotypes: dict is updated in place</span>
<span class="sd">        effects: dict is updated in place</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">epi_rules</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span>
        <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="p">[</span>
            <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="p">[</span><span class="s2">&quot;MUTATION_TYPE&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;EPISTASIS&quot;</span>
        <span class="p">][</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">epi_rules</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># We have some epistasis rules so deal with them</span>
        <span class="n">mutations</span> <span class="o">=</span> <span class="n">subset_multis</span><span class="p">(</span><span class="n">epi_rules</span><span class="p">,</span> <span class="n">mutations</span><span class="p">,</span> <span class="n">just_joined</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">seen_multis</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">effects</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">effects</span><span class="o">.</span><span class="n">keys</span><span class="p">()])</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mutations</span><span class="p">:</span>
            <span class="n">prediction</span> <span class="o">=</span> <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">mutation</span><span class="p">,</span> <span class="n">show_evidence</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">prediction</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                <span class="c1"># prediction == &quot;S&quot; but mypy doesn&#39;t like that</span>
                <span class="c1"># Default prediction so ignore (not that this should happen here)</span>
                <span class="k">continue</span>
            <span class="k">for</span> <span class="n">drug</span> <span class="ow">in</span> <span class="n">prediction</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">pred</span> <span class="o">=</span> <span class="n">prediction</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">pred</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                    <span class="c1"># Shouldn&#39;t be hit but mypy complains</span>
                    <span class="n">pred</span> <span class="o">=</span> <span class="n">pred</span>
                    <span class="n">evidence</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">pred</span><span class="p">,</span> <span class="n">evidence</span> <span class="o">=</span> <span class="n">pred</span>
                <span class="k">if</span> <span class="n">phenotype</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;F&quot;</span><span class="p">:</span>
                    <span class="c1"># F is the only value which overrides epistasis rules</span>
                    <span class="n">phenotype</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="o">=</span> <span class="n">pred</span>
                <span class="c1"># Add to the dict if not already seen</span>
                <span class="c1"># It&#39;s possible that this exact multi already hit an epistasis rule</span>
                <span class="k">if</span> <span class="n">mutation</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">seen_multis</span><span class="p">:</span>
                    <span class="n">effects</span><span class="p">[</span><span class="n">effectsCounter</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
                        <span class="n">vcfStem</span><span class="p">,</span>
                        <span class="kc">None</span><span class="p">,</span>
                        <span class="n">mutation</span><span class="p">,</span>
                        <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                        <span class="n">drug</span><span class="p">,</span>
                        <span class="n">pred</span><span class="p">,</span>
                        <span class="n">evidence</span><span class="p">,</span>
                    <span class="p">]</span>
                    <span class="c1"># Increment counter</span>
                    <span class="n">effectsCounter</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">effectsCounter</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="gnomonicus_lib.getGenes" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">getGenes</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">resistanceCatalogue</span><span class="p">,</span> <span class="n">resistanceGenesOnly</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Get the genes we're interested in.</p>
<p>This is either just resistance genes which have variants, or all which have variants</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>sample</code>
            </td>
            <td>
                  <code><span title="grumpy.Genome">Genome</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Sample's genome object</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>resistanceCatalogue</code>
            </td>
            <td>
                  <code><span title="piezo.ResistanceCatalogue">ResistanceCatalogue</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Resistance catalogue</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>resistanceGenesOnly</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to just use genes within the catalogue</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>set</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>set[str]: Set of gene names</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gnomonicus/gnomonicus_lib.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">getGenes</span><span class="p">(</span>
    <span class="n">sample</span><span class="p">:</span> <span class="n">grumpy</span><span class="o">.</span><span class="n">Genome</span><span class="p">,</span>
    <span class="n">resistanceCatalogue</span><span class="p">:</span> <span class="n">piezo</span><span class="o">.</span><span class="n">ResistanceCatalogue</span><span class="p">,</span>
    <span class="n">resistanceGenesOnly</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">set</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get the genes we&#39;re interested in.</span>

<span class="sd">    This is either just resistance genes which have variants, or all which have variants</span>

<span class="sd">    Args:</span>
<span class="sd">        sample (grumpy.Genome): Sample&#39;s genome object</span>
<span class="sd">        resistanceCatalogue (piezo.ResistanceCatalogue): Resistance catalogue</span>
<span class="sd">        resistanceGenesOnly (bool): Whether to just use genes within the catalogue</span>

<span class="sd">    Returns:</span>
<span class="sd">        set[str]: Set of gene names</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">resistanceCatalogue</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">resistanceGenesOnly</span><span class="p">:</span>
            <span class="n">resistanceGenes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">genes</span><span class="p">)</span>
            <span class="c1"># Catch multi/epistasis rules which might not have specific instances</span>
            <span class="n">multis</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">rule</span> <span class="ow">in</span> <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
                <span class="k">if</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;MUTATION_TYPE&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;MULTI&quot;</span><span class="p">,</span> <span class="s2">&quot;EPISTASIS&quot;</span><span class="p">]:</span>
                    <span class="n">mutations</span> <span class="o">=</span> <span class="n">rule</span><span class="p">[</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">]</span>
                    <span class="k">for</span> <span class="n">mut</span> <span class="ow">in</span> <span class="n">mutations</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&amp;&quot;</span><span class="p">):</span>
                        <span class="n">multis</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">mut</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;@&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
            <span class="n">resistanceGenes</span> <span class="o">=</span> <span class="n">resistanceGenes</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">multis</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">resistanceGenes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">gene_names</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">sample</span><span class="o">.</span><span class="n">genes_with_mutations</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">resistanceGenes</span><span class="p">)</span>

    <span class="k">else</span><span class="p">:</span>
        <span class="c1"># No catalogue, so just stick to genes in the sample</span>
        <span class="k">return</span> <span class="n">sample</span><span class="o">.</span><span class="n">genes</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="gnomonicus_lib.getMutations" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">getMutations</span><span class="p">(</span><span class="n">mutations_df</span><span class="p">,</span> <span class="n">catalogue</span><span class="p">,</span> <span class="n">reference</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Get all of the mutations (including multi-mutations) from the mutations df
Multi-mutations currently only exist within the converted WHO catalogue, and are a highly specific combination
    of mutations which must all be present for a single resistance value.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>mutations_df</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mutations dataframe</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>catalogue</code>
            </td>
            <td>
                  <code><span title="piezo.ResistanceCatalogue">ResistanceCatalogue</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The resistance catalogue. Used to find which multi-mutations we care about</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>reference</code>
            </td>
            <td>
                  <code><span title="grumpy.Genome">Genome</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Reference genome object. Used for checking if mutations are in coding regions</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.List">List</span>[<span title="typing.Tuple">Tuple</span>[str | None, str]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>List[Tuple[str | None, str]]: List of [gene, mutation] or in the case of multi-mutations, [None, multi-mutation]</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gnomonicus/gnomonicus_lib.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span>
<span class="normal">760</span>
<span class="normal">761</span>
<span class="normal">762</span>
<span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span>
<span class="normal">782</span>
<span class="normal">783</span>
<span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span>
<span class="normal">804</span>
<span class="normal">805</span>
<span class="normal">806</span>
<span class="normal">807</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">getMutations</span><span class="p">(</span>
    <span class="n">mutations_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">catalogue</span><span class="p">:</span> <span class="n">piezo</span><span class="o">.</span><span class="n">ResistanceCatalogue</span><span class="p">,</span>
    <span class="n">reference</span><span class="p">:</span> <span class="n">grumpy</span><span class="o">.</span><span class="n">Genome</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get all of the mutations (including multi-mutations) from the mutations df</span>
<span class="sd">    Multi-mutations currently only exist within the converted WHO catalogue, and are a highly specific combination</span>
<span class="sd">        of mutations which must all be present for a single resistance value.</span>

<span class="sd">    Args:</span>
<span class="sd">        mutations_df (pd.DataFrame): Mutations dataframe</span>
<span class="sd">        catalogue (piezo.ResistanceCatalogue): The resistance catalogue. Used to find which multi-mutations we care about</span>
<span class="sd">        reference (grumpy.Genome): Reference genome object. Used for checking if mutations are in coding regions</span>

<span class="sd">    Returns:</span>
<span class="sd">        List[Tuple[str | None, str]]: List of [gene, mutation] or in the case of multi-mutations, [None, multi-mutation]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">mutations_df</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="p">[]</span>
    <span class="n">mutations</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="n">Tuple</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span>
        <span class="nb">zip</span><span class="p">(</span><span class="n">mutations_df</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">],</span> <span class="n">mutations_df</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">])</span>
    <span class="p">)</span>
    <span class="c1"># Grab the multi-mutations from the catalogue</span>
    <span class="c1"># By doing this, we can check a fixed sample space rather than every permutation of the mutations</span>
    <span class="c1"># This makes the problem tractable, but does not address a possible issue with multi-mutations not encapsulating full codons</span>
    <span class="n">multis</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span>
        <span class="n">catalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="p">[</span>
            <span class="n">catalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="p">[</span><span class="s2">&quot;MUTATION_TYPE&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;MULTI&quot;</span>
        <span class="p">][</span><span class="s2">&quot;MUTATION&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">multis</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># We have a catalogue including multi rules, so check if any of these are present in the mutations</span>
        <span class="n">mutations</span> <span class="o">=</span> <span class="n">subset_multis</span><span class="p">(</span><span class="n">multis</span><span class="p">,</span> <span class="n">mutations</span><span class="p">)</span>

    <span class="c1"># Check if the catalogue supports large deletions</span>
    <span class="k">if</span> <span class="s2">&quot;GENE&quot;</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">catalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="p">[</span><span class="s2">&quot;MUTATION_AFFECTS&quot;</span><span class="p">]):</span>
        <span class="n">large_dels</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">large_dels</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="c1"># Filtering out *just* nucleotide changes for cases of synon mutations</span>
    <span class="c1"># The important part of these should have already been found by multi-mutations</span>
    <span class="n">fixed</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mutations</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">reference</span><span class="o">.</span><span class="n">get_gene</span><span class="p">(</span><span class="n">gene</span><span class="p">)</span><span class="o">.</span><span class="n">coding</span><span class="p">:</span>
            <span class="c1"># Codes protein so check for nucleotide changes</span>
            <span class="n">nucleotide</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">                </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                [acgtzx][0-9]+[acgtzx]</span>
<span class="sd">                &quot;&quot;&quot;</span><span class="p">,</span>
                <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">nucleotide</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">mutation</span><span class="p">):</span>
                <span class="c1"># Is a nucleotide (non-promoter) mutation in a coding gene</span>
                <span class="c1"># So skip it as it may cause prediction problems</span>
                <span class="k">continue</span>
        <span class="c1"># Remove large dels if not supported</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">large_dels</span><span class="p">:</span>
            <span class="c1"># Check if this is a large del</span>
            <span class="n">large</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">                </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">                del_(1\.0)|(0\.[0-9][0-9]?[0-9]?)</span>
<span class="sd">                &quot;&quot;&quot;</span><span class="p">,</span>
                <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">large</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">mutation</span><span class="p">):</span>
                <span class="k">continue</span>
        <span class="n">fixed</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">gene</span><span class="p">,</span> <span class="n">mutation</span><span class="p">))</span>
    <span class="k">return</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">fixed</span><span class="p">,</span> <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">x</span><span class="p">]))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="gnomonicus_lib.get_minority_population_type" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_minority_population_type</span><span class="p">(</span><span class="n">catalogue</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Figure out if a catalogue uses FRS or COV. If neither or both, default to FRS</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>catalogue</code>
            </td>
            <td>
                  <code><span title="piezo.ResistanceCatalogue">ResistanceCatalogue</span> | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Catalogue</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="grumpy.MinorType">MinorType</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>grumpy.MinorType: Enum for FRS or COV respectively</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gnomonicus/gnomonicus_lib.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_minority_population_type</span><span class="p">(</span>
    <span class="n">catalogue</span><span class="p">:</span> <span class="n">piezo</span><span class="o">.</span><span class="n">ResistanceCatalogue</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">grumpy</span><span class="o">.</span><span class="n">MinorType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Figure out if a catalogue uses FRS or COV. If neither or both, default to FRS</span>

<span class="sd">    Args:</span>
<span class="sd">        catalogue (piezo.ResistanceCatalogue | None): Catalogue</span>

<span class="sd">    Returns:</span>
<span class="sd">        grumpy.MinorType: Enum for FRS or COV respectively</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">catalogue</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Nothing given, so default to FRS</span>
        <span class="k">return</span> <span class="n">grumpy</span><span class="o">.</span><span class="n">MinorType</span><span class="o">.</span><span class="n">FRS</span>
    <span class="n">frs</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">cov</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">minor</span> <span class="ow">in</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">rules</span><span class="p">[</span><span class="s2">&quot;MINOR&quot;</span><span class="p">]:</span>
        <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">minor</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">m</span><span class="p">:</span>
                <span class="n">m</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>
                <span class="k">assert</span> <span class="n">m</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;Minor populations must be positive: </span><span class="si">{</span><span class="n">m</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="k">if</span> <span class="n">m</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="c1"># FRS</span>
                    <span class="n">frs</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># COV</span>
                    <span class="n">cov</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="c1"># We have just COV</span>
    <span class="k">if</span> <span class="n">cov</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="n">frs</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">grumpy</span><span class="o">.</span><span class="n">MinorType</span><span class="o">.</span><span class="n">COV</span>
    <span class="c1"># We have anything else</span>
    <span class="k">return</span> <span class="n">grumpy</span><span class="o">.</span><span class="n">MinorType</span><span class="o">.</span><span class="n">FRS</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="gnomonicus_lib.parse_grumpy_evidence" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">parse_grumpy_evidence</span><span class="p">(</span><span class="n">evidence</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Parse the grumpy evidence into a dictionary for JSON.</p>
<p>As rust doesn't have multi-type hashmaps, we need to convert all values from strings
Should just be parsing ints and floats</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>evidence</code>
            </td>
            <td>
                  <code><span title="grumpy.VCFRow">VCFRow</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Evidence from the VCF</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>dict</code></td>            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Parsed evidence</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gnomonicus/gnomonicus_lib.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">parse_grumpy_evidence</span><span class="p">(</span><span class="n">evidence</span><span class="p">:</span> <span class="n">grumpy</span><span class="o">.</span><span class="n">VCFRow</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Parse the grumpy evidence into a dictionary for JSON.</span>

<span class="sd">    As rust doesn&#39;t have multi-type hashmaps, we need to convert all values from strings</span>
<span class="sd">    Should just be parsing ints and floats</span>

<span class="sd">    Args:</span>
<span class="sd">        evidence (grumpy.VCFRow): Evidence from the VCF</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: Parsed evidence</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ev</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">evidence</span><span class="o">.</span><span class="n">fields</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">item</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span><span class="p">]</span> <span class="o">|</span> <span class="nb">int</span> <span class="o">|</span> <span class="nb">float</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">if</span> <span class="n">key</span> <span class="o">==</span> <span class="s2">&quot;GT&quot;</span><span class="p">:</span>
            <span class="c1"># Special case here as we need to split the string and parse int/Nones</span>
            <span class="n">gt</span> <span class="o">=</span> <span class="n">value</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)</span>
            <span class="n">item</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">g</span><span class="p">)</span> <span class="k">if</span> <span class="n">g</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;.&quot;</span> <span class="k">else</span> <span class="kc">None</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">gt</span><span class="p">]</span>
        <span class="k">elif</span> <span class="n">key</span> <span class="o">==</span> <span class="s2">&quot;DP&quot;</span><span class="p">:</span>
            <span class="c1"># Single value expected here too (most of the time)</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">value</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="c1"># Odd edge cases around types though</span>
                <span class="k">if</span> <span class="n">value</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span><span class="p">:</span>
                    <span class="n">item</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">item</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">value</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
                    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                        <span class="n">item</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">value</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">item</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">v</span><span class="p">)</span> <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">value</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">item</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">item</span><span class="p">,</span> <span class="nb">int</span><span class="p">)</span> <span class="ow">or</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">item</span><span class="p">,</span> <span class="nb">float</span><span class="p">):</span>
                <span class="c1"># Should never happen but appease mypy</span>
                <span class="k">continue</span>
            <span class="k">for</span> <span class="n">v</span> <span class="ow">in</span> <span class="n">value</span><span class="p">:</span>
                <span class="c1"># Use duck typing to determine if it&#39;s a float or int</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">item</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">v</span><span class="p">))</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">item</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">v</span><span class="p">))</span>
                    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                        <span class="n">item</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">v</span><span class="p">)</span>
        <span class="n">ev</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">item</span>
    <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">val</span> <span class="ow">in</span> <span class="n">ev</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">val</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">val</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
            <span class="c1"># Unpack single values as they probably shouldn&#39;t be lists</span>
            <span class="n">ev</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">val</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># We also want to add back in some of the VCF items which aren&#39;t in the fields dict</span>
    <span class="n">ev</span><span class="p">[</span><span class="s2">&quot;POS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">evidence</span><span class="o">.</span><span class="n">position</span>
    <span class="n">ev</span><span class="p">[</span><span class="s2">&quot;REF&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">evidence</span><span class="o">.</span><span class="n">reference</span>
    <span class="n">ev</span><span class="p">[</span><span class="s2">&quot;ALTS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">evidence</span><span class="o">.</span><span class="n">alternative</span>
    <span class="k">return</span> <span class="n">ev</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="gnomonicus_lib.populateEffects" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">populateEffects</span><span class="p">(</span><span class="n">outputDir</span><span class="p">,</span> <span class="n">resistanceCatalogue</span><span class="p">,</span> <span class="n">mutations</span><span class="p">,</span> <span class="n">vcfStem</span><span class="p">,</span> <span class="n">make_csv</span><span class="p">,</span> <span class="n">make_prediction_csv</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">make_mutations_csv</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">append</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Populate and save the effects DataFrame as a CSV</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>outputDir</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the directory to save the CSV</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>resistanceCatalogue</code>
            </td>
            <td>
                  <code><span title="piezo.ResistanceCatalogue">ResistanceCatalogue</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Resistance catalogue for predictions</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mutations</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mutations dataframe</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>vcfStem</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The basename of the given VCF - used as the sample name</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>make_csv</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to write the CSV of the dataframe</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>make_prediction_csv</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to write the CSV of the antibiogram</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>reference</code>
            </td>
            <td>
                  <code><span title="grumpy.Genome">Genome</span> | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Reference genome. Defaults to None.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>make_mutations_csv</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to write the mutations CSV to disk with new mutations. Defaults to False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>append</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to append data to an existing df at the location (if existing).</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><a class="autorefs autorefs-internal" title="gnomonicus_lib.InvalidMutationException" href="#gnomonicus_lib.InvalidMutationException">InvalidMutationException</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Raised if an invalid mutation is detected</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>(<span title="pandas.DataFrame">DataFrame</span>, dict)</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>(
DataFrame containing the effects data,
A metadata dictionary mapping drugs to their predictions,
DataFrame containing the mutations data</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                  <code><span title="typing.Tuple">Tuple</span>[<span title="pandas.DataFrame">DataFrame</span>, <span title="typing.Dict">Dict</span>, <span title="pandas.DataFrame">DataFrame</span>] | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>)</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gnomonicus/gnomonicus_lib.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 878</span>
<span class="normal"> 879</span>
<span class="normal"> 880</span>
<span class="normal"> 881</span>
<span class="normal"> 882</span>
<span class="normal"> 883</span>
<span class="normal"> 884</span>
<span class="normal"> 885</span>
<span class="normal"> 886</span>
<span class="normal"> 887</span>
<span class="normal"> 888</span>
<span class="normal"> 889</span>
<span class="normal"> 890</span>
<span class="normal"> 891</span>
<span class="normal"> 892</span>
<span class="normal"> 893</span>
<span class="normal"> 894</span>
<span class="normal"> 895</span>
<span class="normal"> 896</span>
<span class="normal"> 897</span>
<span class="normal"> 898</span>
<span class="normal"> 899</span>
<span class="normal"> 900</span>
<span class="normal"> 901</span>
<span class="normal"> 902</span>
<span class="normal"> 903</span>
<span class="normal"> 904</span>
<span class="normal"> 905</span>
<span class="normal"> 906</span>
<span class="normal"> 907</span>
<span class="normal"> 908</span>
<span class="normal"> 909</span>
<span class="normal"> 910</span>
<span class="normal"> 911</span>
<span class="normal"> 912</span>
<span class="normal"> 913</span>
<span class="normal"> 914</span>
<span class="normal"> 915</span>
<span class="normal"> 916</span>
<span class="normal"> 917</span>
<span class="normal"> 918</span>
<span class="normal"> 919</span>
<span class="normal"> 920</span>
<span class="normal"> 921</span>
<span class="normal"> 922</span>
<span class="normal"> 923</span>
<span class="normal"> 924</span>
<span class="normal"> 925</span>
<span class="normal"> 926</span>
<span class="normal"> 927</span>
<span class="normal"> 928</span>
<span class="normal"> 929</span>
<span class="normal"> 930</span>
<span class="normal"> 931</span>
<span class="normal"> 932</span>
<span class="normal"> 933</span>
<span class="normal"> 934</span>
<span class="normal"> 935</span>
<span class="normal"> 936</span>
<span class="normal"> 937</span>
<span class="normal"> 938</span>
<span class="normal"> 939</span>
<span class="normal"> 940</span>
<span class="normal"> 941</span>
<span class="normal"> 942</span>
<span class="normal"> 943</span>
<span class="normal"> 944</span>
<span class="normal"> 945</span>
<span class="normal"> 946</span>
<span class="normal"> 947</span>
<span class="normal"> 948</span>
<span class="normal"> 949</span>
<span class="normal"> 950</span>
<span class="normal"> 951</span>
<span class="normal"> 952</span>
<span class="normal"> 953</span>
<span class="normal"> 954</span>
<span class="normal"> 955</span>
<span class="normal"> 956</span>
<span class="normal"> 957</span>
<span class="normal"> 958</span>
<span class="normal"> 959</span>
<span class="normal"> 960</span>
<span class="normal"> 961</span>
<span class="normal"> 962</span>
<span class="normal"> 963</span>
<span class="normal"> 964</span>
<span class="normal"> 965</span>
<span class="normal"> 966</span>
<span class="normal"> 967</span>
<span class="normal"> 968</span>
<span class="normal"> 969</span>
<span class="normal"> 970</span>
<span class="normal"> 971</span>
<span class="normal"> 972</span>
<span class="normal"> 973</span>
<span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">populateEffects</span><span class="p">(</span>
    <span class="n">outputDir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">resistanceCatalogue</span><span class="p">:</span> <span class="n">piezo</span><span class="o">.</span><span class="n">ResistanceCatalogue</span><span class="p">,</span>
    <span class="n">mutations</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">vcfStem</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">make_csv</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">make_prediction_csv</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">reference</span><span class="p">:</span> <span class="n">grumpy</span><span class="o">.</span><span class="n">Genome</span><span class="p">,</span>
    <span class="n">make_mutations_csv</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">append</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">]</span> <span class="o">|</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Populate and save the effects DataFrame as a CSV</span>

<span class="sd">    Args:</span>
<span class="sd">        outputDir (str): Path to the directory to save the CSV</span>
<span class="sd">        resistanceCatalogue (piezo.ResistanceCatalogue): Resistance catalogue for predictions</span>
<span class="sd">        mutations (pd.DataFrame): Mutations dataframe</span>
<span class="sd">        vcfStem (str): The basename of the given VCF - used as the sample name</span>
<span class="sd">        make_csv (bool): Whether to write the CSV of the dataframe</span>
<span class="sd">        make_prediction_csv (bool): Whether to write the CSV of the antibiogram</span>
<span class="sd">        reference (grumpy.Genome | None, optional): Reference genome. Defaults to None.</span>
<span class="sd">        make_mutations_csv (bool, optional): Whether to write the mutations CSV to disk with new mutations. Defaults to False.</span>
<span class="sd">        append (bool, optional): Whether to append data to an existing df at the location (if existing).</span>

<span class="sd">    Raises:</span>
<span class="sd">        InvalidMutationException: Raised if an invalid mutation is detected</span>

<span class="sd">    Returns:</span>
<span class="sd">        (pd.DataFrame, dict): (</span>
<span class="sd">            DataFrame containing the effects data,</span>
<span class="sd">            A metadata dictionary mapping drugs to their predictions,</span>
<span class="sd">            DataFrame containing the mutations data</span>
<span class="sd">        )</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">resistanceCatalogue</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s2">&quot;Catalogue was None, skipping effects and predictions generation&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="c1"># Assume wildtype behaviour unless otherwise specified</span>
    <span class="n">phenotype</span> <span class="o">=</span> <span class="p">{</span><span class="n">drug</span><span class="p">:</span> <span class="s2">&quot;S&quot;</span> <span class="k">for</span> <span class="n">drug</span> <span class="ow">in</span> <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">drugs</span><span class="p">}</span>

    <span class="n">effects</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">effectsCounter</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1"># Default prediction values are RFUS but use piezo catalogue&#39;s values if existing</span>
    <span class="n">values</span> <span class="o">=</span> <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">values</span>

    <span class="c1"># only try and build an effects table if there are mutations</span>
    <span class="n">sample_mutations</span> <span class="o">=</span> <span class="n">getMutations</span><span class="p">(</span><span class="n">mutations</span><span class="p">,</span> <span class="n">resistanceCatalogue</span><span class="p">,</span> <span class="n">reference</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">sample_mutations</span><span class="p">):</span>
        <span class="c1"># Ensure its a valid mutation</span>
        <span class="c1"># if gene is not None and not referenceGenes[gene].valid_variant(mutation):</span>
        <span class="c1">#     logging.error(f&quot;Not a valid mutation {gene}@{mutation}&quot;)</span>
        <span class="c1">#     raise InvalidMutationException(gene, mutation)</span>

        <span class="c1"># Get the prediction</span>
        <span class="k">if</span> <span class="n">gene</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">prediction</span> <span class="o">=</span> <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span>
                <span class="n">gene</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">mutation</span><span class="p">,</span> <span class="n">show_evidence</span><span class="o">=</span><span class="kc">True</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># This is a multi-mutation so is already of required format</span>
            <span class="n">prediction</span> <span class="o">=</span> <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">mutation</span><span class="p">,</span> <span class="n">show_evidence</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

        <span class="c1"># If the prediction is interesting, iter through drugs to find predictions</span>
        <span class="k">if</span> <span class="n">prediction</span> <span class="o">!=</span> <span class="s2">&quot;S&quot;</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">prediction</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">drug</span> <span class="ow">in</span> <span class="n">prediction</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                <span class="n">drug_pred</span> <span class="o">=</span> <span class="n">prediction</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span>
                <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">drug_pred</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                    <span class="c1"># This shouldn&#39;t happen because we&#39;re showing evidence</span>
                    <span class="c1"># Adding to appease mypy...</span>
                    <span class="n">pred</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">drug_pred</span>
                    <span class="n">evidence</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">pred</span><span class="p">,</span> <span class="n">evidence</span> <span class="o">=</span> <span class="n">drug_pred</span>
                <span class="c1"># Prioritise values based on order within the values list</span>
                <span class="k">if</span> <span class="n">values</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">pred</span><span class="p">)</span> <span class="o">&lt;</span> <span class="n">values</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">phenotype</span><span class="p">[</span><span class="n">drug</span><span class="p">]):</span>
                    <span class="c1"># The prediction is closer to the start of the values list, so should take priority</span>
                    <span class="n">phenotype</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="o">=</span> <span class="n">pred</span>

                <span class="c1"># Add to the dict</span>
                <span class="n">effects</span><span class="p">[</span><span class="n">effectsCounter</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
                    <span class="n">vcfStem</span><span class="p">,</span>
                    <span class="n">gene</span><span class="p">,</span>
                    <span class="n">mutation</span><span class="p">,</span>
                    <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                    <span class="n">drug</span><span class="p">,</span>
                    <span class="n">pred</span><span class="p">,</span>
                    <span class="n">evidence</span><span class="p">,</span>
                <span class="p">]</span>
                <span class="c1"># Increment counter</span>
                <span class="n">effectsCounter</span> <span class="o">+=</span> <span class="mi">1</span>

    <span class="c1"># Check for epistasis rules (which ignore prediction heirarchy)</span>
    <span class="n">effectsCounter</span> <span class="o">=</span> <span class="n">epistasis</span><span class="p">(</span>
        <span class="n">sample_mutations</span><span class="p">,</span>
        <span class="n">resistanceCatalogue</span><span class="p">,</span>
        <span class="n">phenotype</span><span class="p">,</span>
        <span class="n">effects</span><span class="p">,</span>
        <span class="n">effectsCounter</span><span class="p">,</span>
        <span class="n">vcfStem</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Build the DataFrame</span>
    <span class="n">effects_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span>
        <span class="n">effects</span><span class="p">,</span>
        <span class="n">orient</span><span class="o">=</span><span class="s2">&quot;index&quot;</span><span class="p">,</span>
        <span class="n">columns</span><span class="o">=</span><span class="p">[</span>
            <span class="s2">&quot;uniqueid&quot;</span><span class="p">,</span>
            <span class="s2">&quot;gene&quot;</span><span class="p">,</span>
            <span class="s2">&quot;mutation&quot;</span><span class="p">,</span>
            <span class="s2">&quot;catalogue_name&quot;</span><span class="p">,</span>
            <span class="s2">&quot;drug&quot;</span><span class="p">,</span>
            <span class="s2">&quot;prediction&quot;</span><span class="p">,</span>
            <span class="s2">&quot;evidence&quot;</span><span class="p">,</span>
        <span class="p">],</span>
    <span class="p">)</span>
    <span class="n">effects_df</span> <span class="o">=</span> <span class="n">effects_df</span><span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;uniqueid&quot;</span><span class="p">,</span>
            <span class="s2">&quot;gene&quot;</span><span class="p">,</span>
            <span class="s2">&quot;mutation&quot;</span><span class="p">,</span>
            <span class="s2">&quot;drug&quot;</span><span class="p">,</span>
            <span class="s2">&quot;prediction&quot;</span><span class="p">,</span>
            <span class="s2">&quot;catalogue_name&quot;</span><span class="p">,</span>
            <span class="s2">&quot;evidence&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span>
    <span class="n">effects_df</span><span class="p">[</span><span class="s2">&quot;catalogue_version&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">version</span>
    <span class="n">effects_df</span><span class="p">[</span><span class="s2">&quot;prediction_values&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>

    <span class="c1"># Save as CSV</span>
    <span class="k">if</span> <span class="n">make_csv</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">append</span><span class="p">:</span>
            <span class="c1"># Check to see if there&#39;s anything there already</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">old_effects</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outputDir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">vcfStem</span><span class="si">}</span><span class="s2">.effects.csv&quot;</span><span class="p">)</span>
                <span class="p">)</span>
                <span class="n">effects_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">old_effects</span><span class="p">,</span> <span class="n">effects_df</span><span class="p">])</span>
            <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
                <span class="k">pass</span>

        <span class="n">effects_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outputDir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">vcfStem</span><span class="si">}</span><span class="s2">.effects.csv&quot;</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span>
        <span class="p">)</span>

    <span class="n">effects_df</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">make_prediction_csv</span><span class="p">:</span>
        <span class="c1"># We need to construct a simple table here</span>
        <span class="n">predictions</span> <span class="o">=</span> <span class="p">[</span><span class="n">phenotype</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="k">for</span> <span class="n">drug</span> <span class="ow">in</span> <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">drugs</span><span class="p">]</span>
        <span class="n">vals</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;uniqueid&quot;</span><span class="p">:</span> <span class="n">vcfStem</span><span class="p">,</span>
            <span class="s2">&quot;drug&quot;</span><span class="p">:</span> <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">drugs</span><span class="p">,</span>
            <span class="s2">&quot;prediction&quot;</span><span class="p">:</span> <span class="n">predictions</span><span class="p">,</span>
            <span class="s2">&quot;catalogue_name&quot;</span><span class="p">:</span> <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
            <span class="s2">&quot;catalogue_version&quot;</span><span class="p">:</span> <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">version</span><span class="p">,</span>
            <span class="s2">&quot;catalogue_values&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">values</span><span class="p">),</span>
        <span class="p">}</span>
        <span class="n">predictions_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">append</span><span class="p">:</span>
            <span class="c1"># Check to see if there&#39;s anything there already</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">old_predictions</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                    <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outputDir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">vcfStem</span><span class="si">}</span><span class="s2">.predictions.csv&quot;</span><span class="p">)</span>
                <span class="p">)</span>
                <span class="n">predictions_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">old_predictions</span><span class="p">,</span> <span class="n">predictions_df</span><span class="p">])</span>
            <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
                <span class="k">pass</span>
        <span class="n">predictions_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outputDir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">vcfStem</span><span class="si">}</span><span class="s2">.predictions.csv&quot;</span><span class="p">),</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span>
        <span class="p">)</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">effects</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="c1"># We have no effects to report so populate empty df</span>
        <span class="n">effects_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="o">.</span><span class="n">from_dict</span><span class="p">(</span><span class="n">effects</span><span class="p">)</span>

    <span class="c1"># Return  the metadata dict to log later</span>
    <span class="k">return</span> <span class="p">(</span>
        <span class="n">effects_df</span><span class="p">,</span>
        <span class="p">{</span><span class="n">drug</span><span class="p">:</span> <span class="n">phenotype</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="k">for</span> <span class="n">drug</span> <span class="ow">in</span> <span class="n">resistanceCatalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">drugs</span><span class="p">},</span>
        <span class="n">mutations</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="gnomonicus_lib.populateMutations" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">populateMutations</span><span class="p">(</span><span class="n">vcfStem</span><span class="p">,</span> <span class="n">outputDir</span><span class="p">,</span> <span class="n">genome_diff</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">resistanceCatalogue</span><span class="p">,</span> <span class="n">make_csv</span><span class="p">,</span> <span class="n">resistanceGenesOnly</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Popuate and save the mutations DataFrame as a CSV, then return it for use in predictions</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>vcfStem</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The stem of the filename of the VCF file. Used as a uniqueID</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>outputDir</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the desired output directory</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genome_diff</code>
            </td>
            <td>
                  <code><span title="grumpy.GenomeDifference">GenomeDifference</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>GenomeDifference object between reference and this sample</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>reference</code>
            </td>
            <td>
                  <code><span title="grumpy.Genome">Genome</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Reference genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sample</code>
            </td>
            <td>
                  <code><span title="grumpy.Genome">Genome</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Sample genome</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>resistanceCatalogue</code>
            </td>
            <td>
                  <code><span title="piezo.ResistanceCatalogue">ResistanceCatalogue</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Resistance catalogue (used to find which genes to check)</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>make_csv</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to write the CSV of the dataframe</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>resistanceGenesOnly</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to use just genes present in the resistance catalogue</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>MissingFieldException</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Raised when the mutations DataFrame does not contain the required fields</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span> | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: The mutations DataFrame</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gnomonicus/gnomonicus_lib.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">populateMutations</span><span class="p">(</span>
    <span class="n">vcfStem</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">outputDir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">genome_diff</span><span class="p">:</span> <span class="n">grumpy</span><span class="o">.</span><span class="n">GenomeDifference</span><span class="p">,</span>
    <span class="n">reference</span><span class="p">:</span> <span class="n">grumpy</span><span class="o">.</span><span class="n">Genome</span><span class="p">,</span>
    <span class="n">sample</span><span class="p">:</span> <span class="n">grumpy</span><span class="o">.</span><span class="n">Genome</span><span class="p">,</span>
    <span class="n">resistanceCatalogue</span><span class="p">:</span> <span class="n">piezo</span><span class="o">.</span><span class="n">ResistanceCatalogue</span><span class="p">,</span>
    <span class="n">make_csv</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">resistanceGenesOnly</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span> <span class="o">|</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Popuate and save the mutations DataFrame as a CSV, then return it for use in predictions</span>

<span class="sd">    Args:</span>
<span class="sd">        vcfStem (str): The stem of the filename of the VCF file. Used as a uniqueID</span>
<span class="sd">        outputDir (str): Path to the desired output directory</span>
<span class="sd">        genome_diff (grumpy.GenomeDifference): GenomeDifference object between reference and this sample</span>
<span class="sd">        reference (grumpy.Genome): Reference genome</span>
<span class="sd">        sample (grumpy.Genome): Sample genome</span>
<span class="sd">        resistanceCatalogue (piezo.ResistanceCatalogue): Resistance catalogue (used to find which genes to check)</span>
<span class="sd">        make_csv (bool): Whether to write the CSV of the dataframe</span>
<span class="sd">        resistanceGenesOnly (bool): Whether to use just genes present in the resistance catalogue</span>

<span class="sd">    Raises:</span>
<span class="sd">        MissingFieldException: Raised when the mutations DataFrame does not contain the required fields</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: The mutations DataFrame</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genesWithMutations</span> <span class="o">=</span> <span class="n">getGenes</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">resistanceCatalogue</span><span class="p">,</span> <span class="n">resistanceGenesOnly</span><span class="p">)</span>

    <span class="c1"># Iter resistance genes with variation to produce gene level mutations - concating into a single dataframe</span>
    <span class="n">mutations</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;mutation&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;ref&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;alt&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;nucleotide_number&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;nucleotide_index&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;gene_position&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;codes_protein&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;indel_length&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;indel_nucleotides&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;amino_acid_number&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;amino_acid_sequence&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;number_nucleotide_changes&quot;</span><span class="p">:</span> <span class="p">[],</span>
    <span class="p">}</span>
    <span class="k">for</span> <span class="n">gene_name</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">genesWithMutations</span><span class="p">):</span>
        <span class="n">gene_diff</span> <span class="o">=</span> <span class="n">grumpy</span><span class="o">.</span><span class="n">GeneDifference</span><span class="p">(</span>
            <span class="n">reference</span><span class="o">.</span><span class="n">get_gene</span><span class="p">(</span><span class="n">gene_name</span><span class="p">),</span>
            <span class="n">sample</span><span class="o">.</span><span class="n">get_gene</span><span class="p">(</span><span class="n">gene_name</span><span class="p">),</span>
            <span class="n">get_minority_population_type</span><span class="p">(</span><span class="n">resistanceCatalogue</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">gene_diff</span><span class="o">.</span><span class="n">mutations</span><span class="p">:</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">mutation</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">ref_nucleotides</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;alt&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">alt_nucleotides</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;nucleotide_number&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;nucleotide_index&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;gene_position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">gene_position</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;indel_length&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">indel_length</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;indel_nucleotides&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;amino_acid_number&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">amino_acid_number</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;amino_acid_sequence&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">amino_acid_sequence</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;number_nucleotide_changes&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">count_nucleotide_changes</span><span class="p">(</span>
                    <span class="n">mutation</span><span class="o">.</span><span class="n">ref_nucleotides</span><span class="p">,</span> <span class="n">mutation</span><span class="o">.</span><span class="n">alt_nucleotides</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">gene_diff</span><span class="o">.</span><span class="n">minor_mutations</span><span class="p">:</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">mutation</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">ref_nucleotides</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;alt&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">alt_nucleotides</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;nucleotide_number&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">nucleotide_number</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;nucleotide_index&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;gene_position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">gene_position</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">codes_protein</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;indel_length&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">indel_length</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;indel_nucleotides&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;amino_acid_number&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">amino_acid_number</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;amino_acid_sequence&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="o">.</span><span class="n">amino_acid_sequence</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="s2">&quot;number_nucleotide_changes&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">count_nucleotide_changes</span><span class="p">(</span>
                    <span class="n">mutation</span><span class="o">.</span><span class="n">ref_nucleotides</span><span class="p">,</span> <span class="n">mutation</span><span class="o">.</span><span class="n">alt_nucleotides</span>
                <span class="p">)</span>
            <span class="p">)</span>

    <span class="c1"># Ensure correct datatypes</span>
    <span class="n">mutations_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">mutations</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;mutation&quot;</span><span class="p">:</span> <span class="s2">&quot;str&quot;</span><span class="p">,</span>
            <span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="s2">&quot;str&quot;</span><span class="p">,</span>
            <span class="s2">&quot;nucleotide_number&quot;</span><span class="p">:</span> <span class="s2">&quot;Int64&quot;</span><span class="p">,</span>
            <span class="s2">&quot;nucleotide_index&quot;</span><span class="p">:</span> <span class="s2">&quot;Int64&quot;</span><span class="p">,</span>
            <span class="s2">&quot;gene_position&quot;</span><span class="p">:</span> <span class="s2">&quot;Int64&quot;</span><span class="p">,</span>
            <span class="s2">&quot;alt&quot;</span><span class="p">:</span> <span class="s2">&quot;str&quot;</span><span class="p">,</span>
            <span class="s2">&quot;ref&quot;</span><span class="p">:</span> <span class="s2">&quot;str&quot;</span><span class="p">,</span>
            <span class="s2">&quot;codes_protein&quot;</span><span class="p">:</span> <span class="s2">&quot;bool&quot;</span><span class="p">,</span>
            <span class="s2">&quot;indel_length&quot;</span><span class="p">:</span> <span class="s2">&quot;Int64&quot;</span><span class="p">,</span>
            <span class="s2">&quot;indel_nucleotides&quot;</span><span class="p">:</span> <span class="s2">&quot;str&quot;</span><span class="p">,</span>
            <span class="s2">&quot;amino_acid_number&quot;</span><span class="p">:</span> <span class="s2">&quot;Int64&quot;</span><span class="p">,</span>
            <span class="s2">&quot;amino_acid_sequence&quot;</span><span class="p">:</span> <span class="s2">&quot;str&quot;</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="p">)</span>
    <span class="c1"># If there were mutations, write them to a CSV</span>

    <span class="c1"># Add VCF stem as the uniqueID</span>
    <span class="n">mutations_df</span><span class="p">[</span><span class="s2">&quot;uniqueid&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">vcfStem</span>

    <span class="k">if</span> <span class="n">make_csv</span><span class="p">:</span>
        <span class="n">write_mutations_csv</span><span class="p">(</span>
            <span class="n">mutations_df</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outputDir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">vcfStem</span><span class="si">}</span><span class="s2">.mutations.csv&quot;</span><span class="p">)</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="n">mutations_df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="gnomonicus_lib.populateVariants" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">populateVariants</span><span class="p">(</span><span class="n">vcfStem</span><span class="p">,</span> <span class="n">outputDir</span><span class="p">,</span> <span class="n">diff</span><span class="p">,</span> <span class="n">make_csv</span><span class="p">,</span> <span class="n">resistanceGenesOnly</span><span class="p">,</span> <span class="n">sample</span><span class="p">,</span> <span class="n">catalogue</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Populate and save the variants DataFrame as a CSV</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>vcfStem</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>The stem of the filename for the VCF file. Used as a uniqueID</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>outputDir</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the desired output directory</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>diff</code>
            </td>
            <td>
                  <code><span title="grumpy.GenomeDifference">GenomeDifference</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>GenomeDifference object between reference and the sample</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>make_csv</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to write the CSV of the dataframe</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>resistanceGenesOnly</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to use just genes present in the resistance catalogue</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sample</code>
            </td>
            <td>
                  <code><span title="grumpy.Genome">Genome</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Sample genome object</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>catalogue</code>
            </td>
            <td>
                  <code><span title="piezo.ResistanceCatalogue">ResistanceCatalogue</span> | None</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Catalogue for determining FRS or COV for minority populations. If None is given, FRS is assumed. Defaults to None</p>
              </div>
            </td>
            <td>
                  <code>None</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>pd.DataFrame: DataFrame of the variants</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gnomonicus/gnomonicus_lib.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">populateVariants</span><span class="p">(</span>
    <span class="n">vcfStem</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">outputDir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">diff</span><span class="p">:</span> <span class="n">grumpy</span><span class="o">.</span><span class="n">GenomeDifference</span><span class="p">,</span>
    <span class="n">make_csv</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">resistanceGenesOnly</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">sample</span><span class="p">:</span> <span class="n">grumpy</span><span class="o">.</span><span class="n">Genome</span><span class="p">,</span>
    <span class="n">catalogue</span><span class="p">:</span> <span class="n">piezo</span><span class="o">.</span><span class="n">ResistanceCatalogue</span> <span class="o">|</span> <span class="kc">None</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Populate and save the variants DataFrame as a CSV</span>

<span class="sd">    Args:</span>
<span class="sd">        vcfStem (str): The stem of the filename for the VCF file. Used as a uniqueID</span>
<span class="sd">        outputDir (str): Path to the desired output directory</span>
<span class="sd">        diff (grumpy.GenomeDifference): GenomeDifference object between reference and the sample</span>
<span class="sd">        make_csv (bool): Whether to write the CSV of the dataframe</span>
<span class="sd">        resistanceGenesOnly (bool): Whether to use just genes present in the resistance catalogue</span>
<span class="sd">        sample (grumpy.Genome): Sample genome object</span>
<span class="sd">        catalogue (piezo.ResistanceCatalogue | None, optional): Catalogue for determining FRS or COV for minority populations. If None is given, FRS is assumed. Defaults to None</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: DataFrame of the variants</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Populate variants table directly from GenomeDifference</span>
    <span class="n">vals</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">list</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;variant&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;nucleotide_index&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;indel_length&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;indel_nucleotides&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;vcf_evidence&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;vcf_idx&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;gene_position&quot;</span><span class="p">:</span> <span class="p">[],</span>
        <span class="s2">&quot;codon_idx&quot;</span><span class="p">:</span> <span class="p">[],</span>
    <span class="p">}</span>
    <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">diff</span><span class="o">.</span><span class="n">variants</span><span class="p">:</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;variant&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">variant</span><span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;nucleotide_index&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;indel_length&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">indel_length</span><span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;indel_nucleotides&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;vcf_evidence&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">parse_grumpy_evidence</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">get_vcf_row</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">evidence</span><span class="p">)))</span>
        <span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;vcf_idx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">vcf_idx</span><span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">gene_name</span><span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;gene_position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">gene_position</span><span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;codon_idx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">codon_idx</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">diff</span><span class="o">.</span><span class="n">minor_variants</span><span class="p">:</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;variant&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">variant</span><span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;nucleotide_index&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">nucleotide_index</span><span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;indel_length&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">indel_length</span><span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;indel_nucleotides&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">indel_nucleotides</span><span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;vcf_evidence&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
            <span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">parse_grumpy_evidence</span><span class="p">(</span><span class="n">sample</span><span class="o">.</span><span class="n">get_vcf_row</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">evidence</span><span class="p">)))</span>
        <span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;vcf_idx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">vcf_idx</span><span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">gene_name</span><span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;gene_position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">gene_position</span><span class="p">)</span>
        <span class="n">vals</span><span class="p">[</span><span class="s2">&quot;codon_idx&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">variant</span><span class="o">.</span><span class="n">codon_idx</span><span class="p">)</span>

    <span class="c1"># Use of Int64 rather than int is required here as pandas doesn&#39;t allow mixed int/None</span>
    <span class="n">variants</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">vals</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;vcf_evidence&quot;</span><span class="p">:</span> <span class="s2">&quot;object&quot;</span><span class="p">,</span>
            <span class="s2">&quot;nucleotide_index&quot;</span><span class="p">:</span> <span class="s2">&quot;Int64&quot;</span><span class="p">,</span>
            <span class="s2">&quot;indel_length&quot;</span><span class="p">:</span> <span class="s2">&quot;Int64&quot;</span><span class="p">,</span>
            <span class="s2">&quot;vcf_idx&quot;</span><span class="p">:</span> <span class="s2">&quot;Int64&quot;</span><span class="p">,</span>
            <span class="s2">&quot;gene_position&quot;</span><span class="p">:</span> <span class="s2">&quot;Int64&quot;</span><span class="p">,</span>
            <span class="s2">&quot;codon_idx&quot;</span><span class="p">:</span> <span class="s2">&quot;Int64&quot;</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="n">catalogue</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># Figure out if we want to keep all of the variants</span>
        <span class="n">genes</span> <span class="o">=</span> <span class="n">getGenes</span><span class="p">(</span><span class="n">sample</span><span class="p">,</span> <span class="n">catalogue</span><span class="p">,</span> <span class="n">resistanceGenesOnly</span><span class="p">)</span>
        <span class="n">to_drop</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">idx</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
                <span class="c1"># Not a variant we&#39;re interested in, so remove</span>
                <span class="n">to_drop</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx</span><span class="p">)</span>

        <span class="n">variants</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">to_drop</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># Add unique ID to each record</span>
    <span class="n">variants</span><span class="p">[</span><span class="s2">&quot;uniqueid&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">vcfStem</span>

    <span class="n">variants</span> <span class="o">=</span> <span class="n">variants</span><span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;uniqueid&quot;</span><span class="p">,</span>
            <span class="s2">&quot;variant&quot;</span><span class="p">,</span>
            <span class="s2">&quot;gene&quot;</span><span class="p">,</span>
            <span class="s2">&quot;gene_position&quot;</span><span class="p">,</span>
            <span class="s2">&quot;codon_idx&quot;</span><span class="p">,</span>
            <span class="s2">&quot;nucleotide_index&quot;</span><span class="p">,</span>
            <span class="s2">&quot;indel_length&quot;</span><span class="p">,</span>
            <span class="s2">&quot;indel_nucleotides&quot;</span><span class="p">,</span>
            <span class="s2">&quot;vcf_evidence&quot;</span><span class="p">,</span>
            <span class="s2">&quot;vcf_idx&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span>
    <span class="n">variants</span> <span class="o">=</span> <span class="n">variants</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="k">if</span> <span class="n">make_csv</span><span class="p">:</span>
        <span class="c1"># Save CSV</span>
        <span class="n">variants</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
            <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outputDir</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">vcfStem</span><span class="si">}</span><span class="s2">.variants.csv&quot;</span><span class="p">),</span>
            <span class="n">header</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="n">variants</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">variants</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="gnomonicus_lib.saveJSON" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">saveJSON</span><span class="p">(</span><span class="n">variants</span><span class="p">,</span> <span class="n">mutations</span><span class="p">,</span> <span class="n">effects</span><span class="p">,</span> <span class="n">phenotypes</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="n">guid</span><span class="p">,</span> <span class="n">catalogue</span><span class="p">,</span> <span class="n">gnomonicusVersion</span><span class="p">,</span> <span class="n">time_taken</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">vcf_path</span><span class="p">,</span> <span class="n">reference_path</span><span class="p">,</span> <span class="n">catalogue_path</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Create and save a single JSON output file for use within GPAS. JSON structure:
{
    'meta': {
        'status': If this has succeeded or not (but this isn't created in cases it doesn't succeed),
        'workflow_name': 'gnomonicus',
        'workflow_task': 'resistance_prediction' or 'virulenece prediction',
        'workflow_version': gnomonicus.<strong>version</strong>,
        'time_taken': Time this step took,
        'UTC_timestamp': Timestamp for the end of this run,
        'catalogue_type': discrete_values or mic,
        'catalogue_name': Name of catalogue,
        'catalogue_version': Version of the catalogue,
        'reference': Name of the reference genome used,
        'catalogue_file': Path to the catalogue,
        'reference_file': Path to the reference file,
        'vcf_file': Path to the VCF file
    },
    ?'errors': {
        <gene name>: <stack trace>
    }
    'data': {
        'variants': [
            {
                'variant': Genome level variant in GARC,
                'nucleotide_index': Genome index of variant,
                'gene_name': Name of the gene this variant affects (if applicable),
                'gene_position': Gene position which this variant affects. Nucleotide number if non coding, codon indx if coding (if applicable),
                'codon_idx': Index of the base within the corresponding codon this affects (if applicable),
                'vcf_evidence': Parsed VCF row,
                'vcf_idx': Which part of the VCF row to look at for this call
            }, ...
        ],
        ?'mutations': [
            {
                'mutation': Gene level mutation in GARC,
                'gene': Gene name,
                'gene_position': Position within the gene. Amino acid or nucleotide index depending on which is appropriate,
                'vcf_evidence': Parsed VCF row,
                'ref': Ref base(s),
                'alt': Alt base(s)
            }
        ],
        ?'effects': {
            Drug name: [
                {
                    'gene': Gene name of the mutation,
                    'mutation': Gene level mutation in GARC,
                    'prediction': Prediction caused by this mutation,
                    'evidence': Evidence to support this prediction. Currently placeholder
                }, ...,
                {
                    'phenotype': Resultant prediction for this drug based on prediciton heirarchy
                }
            ], ...
        }
        ?'antibiogram': {
            <drug> : <prediction> essentially json[data][effects][<drug>][phenotype]
        }
    }
}
Where fields with a preceeding '?' are not always present depending on data</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>path</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the directory where the variant/mutation/effect CSV files are saved. Also the output dir for this.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>guid</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Sample GUID</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>catalogue</code>
            </td>
            <td>
                  <code><span title="piezo.ResistanceCatalogue">ResistanceCatalogue</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Catalogue used</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gnomonicusVersion</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Semantic versioning string for the gnomonicus module. Can be accessed by <code>gnomonicus.__version__</code></p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>time_taken</code>
            </td>
            <td>
                  <code>float</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Number of seconds taken to run this.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>reference</code>
            </td>
            <td>
                  <code><span title="grumpy.Genome">Genome</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Reference genome object</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>vcf_path</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the VCF file used for this run</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>reference_path</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the reference genome used for this run</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>catalogue_path</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to the catalogue used for this run</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>minor_errors</code>
            </td>
            <td>
                  <code>dict</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Mapping of gene name --&gt; stack trace of any errors occurring when parsing minor mutations</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gnomonicus/gnomonicus_lib.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span>
<span class="normal">1069</span>
<span class="normal">1070</span>
<span class="normal">1071</span>
<span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span>
<span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span>
<span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span>
<span class="normal">1118</span>
<span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span>
<span class="normal">1122</span>
<span class="normal">1123</span>
<span class="normal">1124</span>
<span class="normal">1125</span>
<span class="normal">1126</span>
<span class="normal">1127</span>
<span class="normal">1128</span>
<span class="normal">1129</span>
<span class="normal">1130</span>
<span class="normal">1131</span>
<span class="normal">1132</span>
<span class="normal">1133</span>
<span class="normal">1134</span>
<span class="normal">1135</span>
<span class="normal">1136</span>
<span class="normal">1137</span>
<span class="normal">1138</span>
<span class="normal">1139</span>
<span class="normal">1140</span>
<span class="normal">1141</span>
<span class="normal">1142</span>
<span class="normal">1143</span>
<span class="normal">1144</span>
<span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span>
<span class="normal">1154</span>
<span class="normal">1155</span>
<span class="normal">1156</span>
<span class="normal">1157</span>
<span class="normal">1158</span>
<span class="normal">1159</span>
<span class="normal">1160</span>
<span class="normal">1161</span>
<span class="normal">1162</span>
<span class="normal">1163</span>
<span class="normal">1164</span>
<span class="normal">1165</span>
<span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span>
<span class="normal">1174</span>
<span class="normal">1175</span>
<span class="normal">1176</span>
<span class="normal">1177</span>
<span class="normal">1178</span>
<span class="normal">1179</span>
<span class="normal">1180</span>
<span class="normal">1181</span>
<span class="normal">1182</span>
<span class="normal">1183</span>
<span class="normal">1184</span>
<span class="normal">1185</span>
<span class="normal">1186</span>
<span class="normal">1187</span>
<span class="normal">1188</span>
<span class="normal">1189</span>
<span class="normal">1190</span>
<span class="normal">1191</span>
<span class="normal">1192</span>
<span class="normal">1193</span>
<span class="normal">1194</span>
<span class="normal">1195</span>
<span class="normal">1196</span>
<span class="normal">1197</span>
<span class="normal">1198</span>
<span class="normal">1199</span>
<span class="normal">1200</span>
<span class="normal">1201</span>
<span class="normal">1202</span>
<span class="normal">1203</span>
<span class="normal">1204</span>
<span class="normal">1205</span>
<span class="normal">1206</span>
<span class="normal">1207</span>
<span class="normal">1208</span>
<span class="normal">1209</span>
<span class="normal">1210</span>
<span class="normal">1211</span>
<span class="normal">1212</span>
<span class="normal">1213</span>
<span class="normal">1214</span>
<span class="normal">1215</span>
<span class="normal">1216</span>
<span class="normal">1217</span>
<span class="normal">1218</span>
<span class="normal">1219</span>
<span class="normal">1220</span>
<span class="normal">1221</span>
<span class="normal">1222</span>
<span class="normal">1223</span>
<span class="normal">1224</span>
<span class="normal">1225</span>
<span class="normal">1226</span>
<span class="normal">1227</span>
<span class="normal">1228</span>
<span class="normal">1229</span>
<span class="normal">1230</span>
<span class="normal">1231</span>
<span class="normal">1232</span>
<span class="normal">1233</span>
<span class="normal">1234</span>
<span class="normal">1235</span>
<span class="normal">1236</span>
<span class="normal">1237</span>
<span class="normal">1238</span>
<span class="normal">1239</span>
<span class="normal">1240</span>
<span class="normal">1241</span>
<span class="normal">1242</span>
<span class="normal">1243</span>
<span class="normal">1244</span>
<span class="normal">1245</span>
<span class="normal">1246</span>
<span class="normal">1247</span>
<span class="normal">1248</span>
<span class="normal">1249</span>
<span class="normal">1250</span>
<span class="normal">1251</span>
<span class="normal">1252</span>
<span class="normal">1253</span>
<span class="normal">1254</span>
<span class="normal">1255</span>
<span class="normal">1256</span>
<span class="normal">1257</span>
<span class="normal">1258</span>
<span class="normal">1259</span>
<span class="normal">1260</span>
<span class="normal">1261</span>
<span class="normal">1262</span>
<span class="normal">1263</span>
<span class="normal">1264</span>
<span class="normal">1265</span>
<span class="normal">1266</span>
<span class="normal">1267</span>
<span class="normal">1268</span>
<span class="normal">1269</span>
<span class="normal">1270</span>
<span class="normal">1271</span>
<span class="normal">1272</span>
<span class="normal">1273</span>
<span class="normal">1274</span>
<span class="normal">1275</span>
<span class="normal">1276</span>
<span class="normal">1277</span>
<span class="normal">1278</span>
<span class="normal">1279</span>
<span class="normal">1280</span>
<span class="normal">1281</span>
<span class="normal">1282</span>
<span class="normal">1283</span>
<span class="normal">1284</span>
<span class="normal">1285</span>
<span class="normal">1286</span>
<span class="normal">1287</span>
<span class="normal">1288</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">saveJSON</span><span class="p">(</span>
    <span class="n">variants</span><span class="p">,</span>
    <span class="n">mutations</span><span class="p">,</span>
    <span class="n">effects</span><span class="p">,</span>
    <span class="n">phenotypes</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">str</span><span class="p">],</span>
    <span class="n">path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">guid</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">catalogue</span><span class="p">:</span> <span class="n">piezo</span><span class="o">.</span><span class="n">ResistanceCatalogue</span><span class="p">,</span>
    <span class="n">gnomonicusVersion</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">time_taken</span><span class="p">:</span> <span class="nb">float</span><span class="p">,</span>
    <span class="n">reference</span><span class="p">:</span> <span class="n">grumpy</span><span class="o">.</span><span class="n">Genome</span><span class="p">,</span>
    <span class="n">vcf_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">reference_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">catalogue_path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Create and save a single JSON output file for use within GPAS. JSON structure:</span>
<span class="sd">    {</span>
<span class="sd">        &#39;meta&#39;: {</span>
<span class="sd">            &#39;status&#39;: If this has succeeded or not (but this isn&#39;t created in cases it doesn&#39;t succeed),</span>
<span class="sd">            &#39;workflow_name&#39;: &#39;gnomonicus&#39;,</span>
<span class="sd">            &#39;workflow_task&#39;: &#39;resistance_prediction&#39; or &#39;virulenece prediction&#39;,</span>
<span class="sd">            &#39;workflow_version&#39;: gnomonicus.__version__,</span>
<span class="sd">            &#39;time_taken&#39;: Time this step took,</span>
<span class="sd">            &#39;UTC_timestamp&#39;: Timestamp for the end of this run,</span>
<span class="sd">            &#39;catalogue_type&#39;: discrete_values or mic,</span>
<span class="sd">            &#39;catalogue_name&#39;: Name of catalogue,</span>
<span class="sd">            &#39;catalogue_version&#39;: Version of the catalogue,</span>
<span class="sd">            &#39;reference&#39;: Name of the reference genome used,</span>
<span class="sd">            &#39;catalogue_file&#39;: Path to the catalogue,</span>
<span class="sd">            &#39;reference_file&#39;: Path to the reference file,</span>
<span class="sd">            &#39;vcf_file&#39;: Path to the VCF file</span>
<span class="sd">        },</span>
<span class="sd">        ?&#39;errors&#39;: {</span>
<span class="sd">            &lt;gene name&gt;: &lt;stack trace&gt;</span>
<span class="sd">        }</span>
<span class="sd">        &#39;data&#39;: {</span>
<span class="sd">            &#39;variants&#39;: [</span>
<span class="sd">                {</span>
<span class="sd">                    &#39;variant&#39;: Genome level variant in GARC,</span>
<span class="sd">                    &#39;nucleotide_index&#39;: Genome index of variant,</span>
<span class="sd">                    &#39;gene_name&#39;: Name of the gene this variant affects (if applicable),</span>
<span class="sd">                    &#39;gene_position&#39;: Gene position which this variant affects. Nucleotide number if non coding, codon indx if coding (if applicable),</span>
<span class="sd">                    &#39;codon_idx&#39;: Index of the base within the corresponding codon this affects (if applicable),</span>
<span class="sd">                    &#39;vcf_evidence&#39;: Parsed VCF row,</span>
<span class="sd">                    &#39;vcf_idx&#39;: Which part of the VCF row to look at for this call</span>
<span class="sd">                }, ...</span>
<span class="sd">            ],</span>
<span class="sd">            ?&#39;mutations&#39;: [</span>
<span class="sd">                {</span>
<span class="sd">                    &#39;mutation&#39;: Gene level mutation in GARC,</span>
<span class="sd">                    &#39;gene&#39;: Gene name,</span>
<span class="sd">                    &#39;gene_position&#39;: Position within the gene. Amino acid or nucleotide index depending on which is appropriate,</span>
<span class="sd">                    &#39;vcf_evidence&#39;: Parsed VCF row,</span>
<span class="sd">                    &#39;ref&#39;: Ref base(s),</span>
<span class="sd">                    &#39;alt&#39;: Alt base(s)</span>
<span class="sd">                }</span>
<span class="sd">            ],</span>
<span class="sd">            ?&#39;effects&#39;: {</span>
<span class="sd">                Drug name: [</span>
<span class="sd">                    {</span>
<span class="sd">                        &#39;gene&#39;: Gene name of the mutation,</span>
<span class="sd">                        &#39;mutation&#39;: Gene level mutation in GARC,</span>
<span class="sd">                        &#39;prediction&#39;: Prediction caused by this mutation,</span>
<span class="sd">                        &#39;evidence&#39;: Evidence to support this prediction. Currently placeholder</span>
<span class="sd">                    }, ...,</span>
<span class="sd">                    {</span>
<span class="sd">                        &#39;phenotype&#39;: Resultant prediction for this drug based on prediciton heirarchy</span>
<span class="sd">                    }</span>
<span class="sd">                ], ...</span>
<span class="sd">            }</span>
<span class="sd">            ?&#39;antibiogram&#39;: {</span>
<span class="sd">                &lt;drug&gt; : &lt;prediction&gt; essentially json[data][effects][&lt;drug&gt;][phenotype]</span>
<span class="sd">            }</span>
<span class="sd">        }</span>
<span class="sd">    }</span>
<span class="sd">    Where fields with a preceeding &#39;?&#39; are not always present depending on data</span>

<span class="sd">    Args:</span>
<span class="sd">        path (str): Path to the directory where the variant/mutation/effect CSV files are saved. Also the output dir for this.</span>
<span class="sd">        guid (str): Sample GUID</span>
<span class="sd">        catalogue (piezo.ResistanceCatalogue): Catalogue used</span>
<span class="sd">        gnomonicusVersion (str): Semantic versioning string for the gnomonicus module. Can be accessed by `gnomonicus.__version__`</span>
<span class="sd">        time_taken (float): Number of seconds taken to run this.</span>
<span class="sd">        reference (grumpy.Genome): Reference genome object</span>
<span class="sd">        vcf_path (str): Path to the VCF file used for this run</span>
<span class="sd">        reference_path (str): Path to the reference genome used for this run</span>
<span class="sd">        catalogue_path (str): Path to the catalogue used for this run</span>
<span class="sd">        minor_errors (dict): Mapping of gene name --&gt; stack trace of any errors occurring when parsing minor mutations</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Define some metadata for the json</span>
    <span class="n">meta</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span>
        <span class="p">{</span>
            <span class="s2">&quot;status&quot;</span><span class="p">:</span> <span class="s2">&quot;success&quot;</span><span class="p">,</span>
            <span class="s2">&quot;workflow_name&quot;</span><span class="p">:</span> <span class="s2">&quot;gnomonicus&quot;</span><span class="p">,</span>
            <span class="s2">&quot;workflow_version&quot;</span><span class="p">:</span> <span class="n">gnomonicusVersion</span><span class="p">,</span>  <span class="c1"># gnomonicus version used</span>
            <span class="s2">&quot;workflow_task&quot;</span><span class="p">:</span> <span class="s2">&quot;resistance_prediction&quot;</span><span class="p">,</span>  <span class="c1"># TODO: Update this when we know how to detect a virulence catalogue</span>
            <span class="s2">&quot;guid&quot;</span><span class="p">:</span> <span class="n">guid</span><span class="p">,</span>  <span class="c1"># Sample GUID</span>
            <span class="s2">&quot;UTC-datetime-completed&quot;</span><span class="p">:</span> <span class="n">datetime</span><span class="o">.</span><span class="n">datetime</span><span class="o">.</span><span class="n">utcnow</span><span class="p">()</span><span class="o">.</span><span class="n">isoformat</span><span class="p">(),</span>  <span class="c1"># ISO datetime run</span>
            <span class="s2">&quot;time_taken_s&quot;</span><span class="p">:</span> <span class="n">time_taken</span><span class="p">,</span>
            <span class="s2">&quot;reference&quot;</span><span class="p">:</span> <span class="n">reference</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
            <span class="s2">&quot;catalogue_file&quot;</span><span class="p">:</span> <span class="n">catalogue_path</span><span class="p">,</span>
            <span class="s2">&quot;reference_file&quot;</span><span class="p">:</span> <span class="n">reference_path</span><span class="p">,</span>
            <span class="s2">&quot;vcf_file&quot;</span><span class="p">:</span> <span class="n">vcf_path</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="n">catalogue</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">meta</span><span class="p">[</span><span class="s2">&quot;catalogue_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">catalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
        <span class="n">meta</span><span class="p">[</span><span class="s2">&quot;catalogue_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">name</span>
        <span class="n">meta</span><span class="p">[</span><span class="s2">&quot;catalogue_version&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">catalogue</span><span class="o">.</span><span class="n">catalogue</span><span class="o">.</span><span class="n">version</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">meta</span><span class="p">[</span><span class="s2">&quot;catalogue_type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">meta</span><span class="p">[</span><span class="s2">&quot;catalogue_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">meta</span><span class="p">[</span><span class="s2">&quot;catalogue_version&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>

    <span class="c1"># Main data collection</span>
    <span class="n">data</span><span class="p">:</span> <span class="n">Dict</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>

    <span class="c1"># Antibigram field</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;antibiogram&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span>
        <span class="p">[(</span><span class="n">key</span><span class="p">,</span> <span class="n">phenotypes</span><span class="p">[</span><span class="n">key</span><span class="p">])</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">phenotypes</span><span class="o">.</span><span class="n">keys</span><span class="p">())]</span>
    <span class="p">)</span>

    <span class="c1"># Variants field</span>
    <span class="n">_variants</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">variants</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
        <span class="n">row</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span>
            <span class="p">{</span>
                <span class="s2">&quot;variant&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="n">variant</span><span class="p">[</span><span class="s2">&quot;variant&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">variant</span><span class="p">[</span><span class="s2">&quot;variant&quot;</span><span class="p">])</span> <span class="k">else</span> <span class="kc">None</span>
                <span class="p">),</span>
                <span class="s2">&quot;nucleotide_index&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="n">variant</span><span class="p">[</span><span class="s2">&quot;nucleotide_index&quot;</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">variant</span><span class="p">[</span><span class="s2">&quot;nucleotide_index&quot;</span><span class="p">])</span>
                    <span class="k">else</span> <span class="kc">None</span>
                <span class="p">),</span>
                <span class="s2">&quot;gene_name&quot;</span><span class="p">:</span> <span class="n">variant</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">variant</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">])</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                <span class="s2">&quot;gene_position&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="n">variant</span><span class="p">[</span><span class="s2">&quot;gene_position&quot;</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">variant</span><span class="p">[</span><span class="s2">&quot;gene_position&quot;</span><span class="p">])</span>
                    <span class="k">else</span> <span class="kc">None</span>
                <span class="p">),</span>
                <span class="s2">&quot;codon_idx&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="n">variant</span><span class="p">[</span><span class="s2">&quot;codon_idx&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">variant</span><span class="p">[</span><span class="s2">&quot;codon_idx&quot;</span><span class="p">])</span> <span class="k">else</span> <span class="kc">None</span>
                <span class="p">),</span>
                <span class="s2">&quot;vcf_evidence&quot;</span><span class="p">:</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">variant</span><span class="p">[</span><span class="s2">&quot;vcf_evidence&quot;</span><span class="p">]),</span>
                <span class="s2">&quot;vcf_idx&quot;</span><span class="p">:</span> <span class="p">(</span>
                    <span class="n">variant</span><span class="p">[</span><span class="s2">&quot;vcf_idx&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">variant</span><span class="p">[</span><span class="s2">&quot;vcf_idx&quot;</span><span class="p">])</span> <span class="k">else</span> <span class="kc">None</span>
                <span class="p">),</span>
            <span class="p">}</span>
        <span class="p">)</span>
        <span class="n">_variants</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>
    <span class="n">_variants</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
        <span class="n">_variants</span><span class="p">,</span>
        <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span>
            <span class="n">x</span><span class="p">[</span><span class="s2">&quot;gene_name&quot;</span><span class="p">]</span> <span class="ow">or</span> <span class="s2">&quot;z&quot;</span><span class="p">,</span>
            <span class="n">x</span><span class="p">[</span><span class="s2">&quot;gene_position&quot;</span><span class="p">]</span> <span class="ow">or</span> <span class="mi">0</span><span class="p">,</span>
            <span class="n">x</span><span class="p">[</span><span class="s2">&quot;variant&quot;</span><span class="p">]</span> <span class="ow">or</span> <span class="s2">&quot;z&quot;</span><span class="p">,</span>
        <span class="p">),</span>
    <span class="p">)</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;variants&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_variants</span>

    <span class="c1"># Depending on mutations/effects, populate</span>
    <span class="n">_mutations</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">if</span> <span class="n">mutations</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">mutations</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">row</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="n">mutation</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">mutation</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">])</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="s2">&quot;gene_position&quot;</span><span class="p">:</span> <span class="p">(</span>
                        <span class="n">mutation</span><span class="p">[</span><span class="s2">&quot;gene_position&quot;</span><span class="p">]</span>
                        <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">mutation</span><span class="p">[</span><span class="s2">&quot;gene_position&quot;</span><span class="p">])</span>
                        <span class="k">else</span> <span class="kc">None</span>
                    <span class="p">),</span>
                    <span class="s2">&quot;mutation&quot;</span><span class="p">:</span> <span class="p">(</span>
                        <span class="n">mutation</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">]</span>
                        <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">mutation</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">])</span>
                        <span class="k">else</span> <span class="kc">None</span>
                    <span class="p">),</span>
                <span class="p">}</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">mutation</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">isupper</span><span class="p">()</span> <span class="ow">or</span> <span class="n">mutation</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;!&quot;</span><span class="p">:</span>
                <span class="c1"># Only add codon ref/alt for AA changes</span>
                <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mutation</span><span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">mutation</span><span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">])</span> <span class="k">else</span> <span class="kc">None</span>
                <span class="n">row</span><span class="p">[</span><span class="s2">&quot;alt&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mutation</span><span class="p">[</span><span class="s2">&quot;alt&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">mutation</span><span class="p">[</span><span class="s2">&quot;alt&quot;</span><span class="p">])</span> <span class="k">else</span> <span class="kc">None</span>
            <span class="n">_mutations</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">row</span><span class="p">)</span>

    <span class="n">_mutations</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
        <span class="n">_mutations</span><span class="p">,</span>
        <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span> <span class="ow">or</span> <span class="s2">&quot;z&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;gene_position&quot;</span><span class="p">]</span> <span class="ow">or</span> <span class="mi">0</span><span class="p">),</span>
    <span class="p">)</span>
    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;mutations&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_mutations</span>

    <span class="n">_effects</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="nb">list</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">effects</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">effects</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">effect</span> <span class="ow">in</span> <span class="n">effects</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="n">prediction</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span>
                <span class="p">{</span>
                    <span class="s2">&quot;gene&quot;</span><span class="p">:</span> <span class="n">effect</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">effect</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">])</span> <span class="k">else</span> <span class="kc">None</span><span class="p">,</span>
                    <span class="s2">&quot;mutation&quot;</span><span class="p">:</span> <span class="p">(</span>
                        <span class="n">effect</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">effect</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">])</span> <span class="k">else</span> <span class="kc">None</span>
                    <span class="p">),</span>
                    <span class="s2">&quot;prediction&quot;</span><span class="p">:</span> <span class="p">(</span>
                        <span class="n">effect</span><span class="p">[</span><span class="s2">&quot;prediction&quot;</span><span class="p">]</span>
                        <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">notnull</span><span class="p">(</span><span class="n">effect</span><span class="p">[</span><span class="s2">&quot;prediction&quot;</span><span class="p">])</span>
                        <span class="k">else</span> <span class="kc">None</span>
                    <span class="p">),</span>
                    <span class="s2">&quot;evidence&quot;</span><span class="p">:</span> <span class="n">effect</span><span class="p">[</span><span class="s2">&quot;evidence&quot;</span><span class="p">],</span>
                <span class="p">}</span>
            <span class="p">)</span>
            <span class="n">_effects</span><span class="p">[</span><span class="n">effect</span><span class="p">[</span><span class="s2">&quot;drug&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prediction</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">drug</span> <span class="ow">in</span> <span class="n">_effects</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
        <span class="n">_effects</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span>
            <span class="n">_effects</span><span class="p">[</span><span class="n">drug</span><span class="p">],</span>
            <span class="n">key</span><span class="o">=</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span><span class="p">]</span> <span class="ow">or</span> <span class="s2">&quot;z&quot;</span><span class="p">,</span> <span class="n">x</span><span class="p">[</span><span class="s2">&quot;mutation&quot;</span><span class="p">]</span> <span class="ow">or</span> <span class="s2">&quot;z&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="n">_effects</span><span class="p">[</span><span class="n">drug</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">OrderedDict</span><span class="p">({</span><span class="s2">&quot;phenotype&quot;</span><span class="p">:</span> <span class="n">phenotypes</span><span class="p">[</span><span class="n">drug</span><span class="p">]}))</span>

    <span class="n">data</span><span class="p">[</span><span class="s2">&quot;effects&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">(</span>
        <span class="p">[(</span><span class="n">key</span><span class="p">,</span> <span class="n">_effects</span><span class="p">[</span><span class="n">key</span><span class="p">])</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">_effects</span><span class="o">.</span><span class="n">keys</span><span class="p">())]</span>
    <span class="p">)</span>

    <span class="c1"># Convert fields to a list so it can be json serialised</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">guid</span><span class="si">}</span><span class="s2">.gnomonicus-out.json&quot;</span><span class="p">),</span> <span class="s2">&quot;w&quot;</span><span class="p">,</span> <span class="n">encoding</span><span class="o">=</span><span class="s2">&quot;utf-8&quot;</span>
    <span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">({</span><span class="s2">&quot;meta&quot;</span><span class="p">:</span> <span class="n">meta</span><span class="p">,</span> <span class="s2">&quot;data&quot;</span><span class="p">:</span> <span class="n">data</span><span class="p">},</span> <span class="n">indent</span><span class="o">=</span><span class="mi">2</span><span class="p">))</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="gnomonicus_lib.subset_multis" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">subset_multis</span><span class="p">(</span><span class="n">multis</span><span class="p">,</span> <span class="n">mutations_</span><span class="p">,</span> <span class="n">just_joined</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Given a list of multis, combine existing mutations to match them</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>multis</code>
            </td>
            <td>
                  <code>set[str]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Set of multi mutations</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>mutations_</code>
            </td>
            <td>
                  <code>list[tuple[str | None, str]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>mutations in</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>just_joined</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to return all mutations or just joined multis. Defaults to False</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>list[tuple[str | None, str]]</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>list[tuple[str|None, str]]: Multi-mutations which should hit catalogue rules</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gnomonicus/gnomonicus_lib.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span>
<span class="normal">722</span>
<span class="normal">723</span>
<span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">subset_multis</span><span class="p">(</span>
    <span class="n">multis</span><span class="p">:</span> <span class="nb">set</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">mutations_</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span> <span class="nb">str</span><span class="p">]],</span>
    <span class="n">just_joined</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Given a list of multis, combine existing mutations to match them</span>

<span class="sd">    Args:</span>
<span class="sd">        multis (set[str]): Set of multi mutations</span>
<span class="sd">        mutations_ (list[tuple[str | None, str]]): mutations in</span>
<span class="sd">        just_joined (bool, optional): Whether to return all mutations or just joined multis. Defaults to False</span>

<span class="sd">    Returns:</span>
<span class="sd">        list[tuple[str|None, str]]: Multi-mutations which should hit catalogue rules</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mutations</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">mut</span><span class="p">,</span> <span class="n">mut</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="s2">&quot;:&quot;</span> <span class="ow">in</span> <span class="n">mut</span> <span class="k">else</span> <span class="kc">None</span><span class="p">)</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">mut</span><span class="p">)</span> <span class="ow">in</span> <span class="n">mutations_</span>
        <span class="k">if</span> <span class="n">gene</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="kc">None</span>
    <span class="p">]</span>
    <span class="n">joined</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">gene</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">mut</span> <span class="k">for</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">mut</span><span class="p">,</span> <span class="n">_</span><span class="p">)</span> <span class="ow">in</span> <span class="n">mutations</span><span class="p">])</span>
    <span class="n">large_del_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        .*del_[01]\.[0-9]+.*</span>
<span class="sd">        &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">snp_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        [acgtxzA-Z!]-?[0-9]+[acgtxzA-Z!]</span>
<span class="sd">    &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">early_stop_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        [A-Z!]-?[0-9]+!</span>
<span class="sd">    &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">indel_re</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">compile</span><span class="p">(</span>
<span class="w">        </span><span class="sa">r</span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        -?[0-9]+_(ins|del|indel|mixed|fs)(_(([0-9]+)|([a-z]+)))?</span>
<span class="sd">    &quot;&quot;&quot;</span><span class="p">,</span>
        <span class="n">re</span><span class="o">.</span><span class="n">VERBOSE</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Find these once so they aren&#39;t fetched on every iteration</span>
    <span class="n">existing_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">([</span><span class="n">gene</span> <span class="k">for</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">)</span> <span class="ow">in</span> <span class="n">mutations</span><span class="p">])</span>
    <span class="n">large_dels</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">mut</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">mut</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span> <span class="ow">in</span> <span class="n">mutations</span>
        <span class="k">if</span> <span class="n">large_del_re</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">mut</span><span class="p">)</span>
    <span class="p">]</span>
    <span class="n">snps</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">mut</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span> <span class="k">for</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">mut</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span> <span class="ow">in</span> <span class="n">mutations</span> <span class="k">if</span> <span class="n">snp_re</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">mut</span><span class="p">)</span>
    <span class="p">]</span>
    <span class="n">early_stop</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">mut</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">mut</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span> <span class="ow">in</span> <span class="n">mutations</span>
        <span class="k">if</span> <span class="n">early_stop_re</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">mut</span><span class="p">)</span>
    <span class="p">]</span>
    <span class="n">indels</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">mut</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span>
        <span class="k">for</span> <span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">mut</span><span class="p">,</span> <span class="n">minor</span><span class="p">)</span> <span class="ow">in</span> <span class="n">mutations</span>
        <span class="k">if</span> <span class="n">indel_re</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">mut</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="n">new_mutations</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">str</span> <span class="o">|</span> <span class="kc">None</span><span class="p">,</span> <span class="nb">str</span><span class="p">]]</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">multi</span> <span class="ow">in</span> <span class="n">multis</span><span class="p">:</span>
        <span class="n">multi_match</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">check</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">if</span> <span class="s2">&quot;*&quot;</span> <span class="ow">in</span> <span class="n">multi</span> <span class="ow">or</span> <span class="s2">&quot;?&quot;</span> <span class="ow">in</span> <span class="n">multi</span> <span class="ow">or</span> <span class="n">large_del_re</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">multi</span><span class="p">):</span>
            <span class="c1"># We need to be a bit more cleaver here to avoid combinatorics ruining things</span>
            <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">multi</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&amp;&quot;</span><span class="p">):</span>
                <span class="n">gene</span><span class="p">,</span> <span class="n">mut</span> <span class="o">=</span> <span class="n">mutation</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;@&quot;</span><span class="p">)</span>
                <span class="n">rule_is_minor</span> <span class="o">=</span> <span class="s2">&quot;:&quot;</span> <span class="ow">in</span> <span class="n">mut</span>
                <span class="n">this_match</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">if</span> <span class="n">gene</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">existing_genes</span><span class="p">:</span>
                    <span class="c1"># Gene doesn&#39;t exist in this sample, so skip</span>
                    <span class="n">check</span> <span class="o">=</span> <span class="kc">False</span>
                <span class="k">if</span> <span class="s2">&quot;*&quot;</span> <span class="ow">in</span> <span class="n">mutation</span><span class="p">:</span>
                    <span class="c1"># A few cases here, wildcard or early-stop SNP, ins, del, indel, fs</span>
                    <span class="k">if</span> <span class="s2">&quot;?&quot;</span> <span class="ow">in</span> <span class="n">mutation</span><span class="p">:</span>
                        <span class="c1"># Wildcard-non-synon SNP so match any SNP in this gene</span>
                        <span class="k">if</span> <span class="s2">&quot;-&quot;</span> <span class="ow">in</span> <span class="n">mutation</span><span class="p">:</span>
                            <span class="n">promoter</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">promoter</span> <span class="o">=</span> <span class="kc">False</span>
                        <span class="n">matched</span> <span class="o">=</span> <span class="kc">False</span>
                        <span class="k">for</span> <span class="n">g</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">minor</span> <span class="ow">in</span> <span class="n">snps</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">g</span> <span class="o">==</span> <span class="n">gene</span> <span class="ow">and</span> <span class="p">(</span>
                                <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                                <span class="ow">or</span> <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                            <span class="p">):</span>
                                <span class="k">if</span> <span class="n">promoter</span> <span class="ow">and</span> <span class="s2">&quot;-&quot;</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
                                    <span class="n">matched</span> <span class="o">=</span> <span class="kc">True</span>
                                    <span class="k">if</span> <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                                        <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">minor</span><span class="p">)</span>
                                    <span class="k">else</span><span class="p">:</span>
                                        <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span><span class="p">)</span>
                                <span class="k">elif</span> <span class="ow">not</span> <span class="n">promoter</span> <span class="ow">and</span> <span class="s2">&quot;-&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
                                    <span class="n">matched</span> <span class="o">=</span> <span class="kc">True</span>
                                    <span class="k">if</span> <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                                        <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">minor</span><span class="p">)</span>
                                    <span class="k">else</span><span class="p">:</span>
                                        <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span><span class="p">)</span>
                        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">matched</span>
                    <span class="k">elif</span> <span class="s2">&quot;!&quot;</span> <span class="ow">in</span> <span class="n">mutation</span><span class="p">:</span>
                        <span class="n">matched</span> <span class="o">=</span> <span class="kc">False</span>
                        <span class="k">for</span> <span class="n">g</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">minor</span> <span class="ow">in</span> <span class="n">early_stop</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">g</span> <span class="o">==</span> <span class="n">gene</span> <span class="ow">and</span> <span class="p">(</span>
                                <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                                <span class="ow">or</span> <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                            <span class="p">):</span>
                                <span class="n">matched</span> <span class="o">=</span> <span class="kc">True</span>
                                <span class="k">if</span> <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                                    <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">minor</span><span class="p">)</span>
                                <span class="k">else</span><span class="p">:</span>
                                    <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span><span class="p">)</span>
                        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">matched</span>
                    <span class="k">elif</span> <span class="s2">&quot;=&quot;</span> <span class="ow">in</span> <span class="n">mutation</span><span class="p">:</span>
                        <span class="n">matched</span> <span class="o">=</span> <span class="kc">False</span>
                        <span class="k">for</span> <span class="n">g</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">minor</span> <span class="ow">in</span> <span class="n">snps</span><span class="p">:</span>
                            <span class="k">if</span> <span class="p">(</span>
                                <span class="n">g</span> <span class="o">==</span> <span class="n">gene</span>
                                <span class="ow">and</span> <span class="n">m</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">m</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                                <span class="ow">and</span> <span class="p">(</span>
                                    <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                                    <span class="ow">or</span> <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                                <span class="p">)</span>
                            <span class="p">):</span>
                                <span class="n">matched</span> <span class="o">=</span> <span class="kc">True</span>
                                <span class="k">if</span> <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                                    <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">minor</span><span class="p">)</span>
                                <span class="k">else</span><span class="p">:</span>
                                    <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span><span class="p">)</span>
                        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">matched</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="c1"># Wildcard indels never have associated numbers or bases (as that wouldn&#39;t make sense)</span>
                        <span class="c1"># So check for matches</span>
                        <span class="n">matched</span> <span class="o">=</span> <span class="kc">False</span>
                        <span class="k">if</span> <span class="s2">&quot;-&quot;</span> <span class="ow">in</span> <span class="n">mutation</span><span class="p">:</span>
                            <span class="n">promoter</span> <span class="o">=</span> <span class="kc">True</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">promoter</span> <span class="o">=</span> <span class="kc">False</span>

                        <span class="k">if</span> <span class="s2">&quot;ins&quot;</span> <span class="ow">in</span> <span class="n">mutation</span> <span class="ow">or</span> <span class="s2">&quot;del&quot;</span> <span class="ow">in</span> <span class="n">mutation</span><span class="p">:</span>
                            <span class="k">if</span> <span class="s2">&quot;ins&quot;</span> <span class="ow">in</span> <span class="n">mutation</span><span class="p">:</span>
                                <span class="n">searching</span> <span class="o">=</span> <span class="s2">&quot;ins&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">searching</span> <span class="o">=</span> <span class="s2">&quot;del&quot;</span>
                            <span class="k">for</span> <span class="n">g</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">minor</span> <span class="ow">in</span> <span class="n">indels</span><span class="p">:</span>
                                <span class="k">if</span> <span class="n">g</span> <span class="o">==</span> <span class="n">gene</span> <span class="ow">and</span> <span class="p">(</span>
                                    <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                                    <span class="ow">or</span> <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                                <span class="p">):</span>
                                    <span class="k">if</span> <span class="n">promoter</span> <span class="ow">and</span> <span class="s2">&quot;-&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
                                        <span class="k">continue</span>
                                    <span class="k">elif</span> <span class="ow">not</span> <span class="n">promoter</span> <span class="ow">and</span> <span class="s2">&quot;-&quot;</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
                                        <span class="k">continue</span>
                                    <span class="k">if</span> <span class="n">searching</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
                                        <span class="n">matched</span> <span class="o">=</span> <span class="kc">True</span>
                                        <span class="k">if</span> <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                                            <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">minor</span><span class="p">)</span>
                                        <span class="k">else</span><span class="p">:</span>
                                            <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span><span class="p">)</span>
                        <span class="k">elif</span> <span class="s2">&quot;fs&quot;</span> <span class="ow">in</span> <span class="n">mutation</span><span class="p">:</span>
                            <span class="c1"># Bit more annoying here as we have to check if specific indels are framshifting</span>
                            <span class="k">for</span> <span class="n">g</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">minor</span> <span class="ow">in</span> <span class="n">indels</span><span class="p">:</span>
                                <span class="k">if</span> <span class="n">g</span> <span class="o">==</span> <span class="n">gene</span> <span class="ow">and</span> <span class="p">(</span>
                                    <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                                    <span class="ow">or</span> <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                                <span class="p">):</span>
                                    <span class="c1"># Don&#39;t need to check for promoters here as promoter framshift doesn&#39;t make sense</span>
                                    <span class="n">bases_in</span> <span class="o">=</span> <span class="n">m</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                                    <span class="k">if</span> <span class="n">bases_in</span><span class="o">.</span><span class="n">isnumeric</span><span class="p">():</span>
                                        <span class="c1"># Number of bases rather than actual bases</span>
                                        <span class="n">bases</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">bases_in</span><span class="p">)</span>
                                    <span class="k">else</span><span class="p">:</span>
                                        <span class="n">bases</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">bases_in</span><span class="p">)</span>
                                    <span class="k">if</span> <span class="n">bases</span> <span class="o">%</span> <span class="mi">3</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                                        <span class="n">matched</span> <span class="o">=</span> <span class="kc">True</span>
                                        <span class="k">if</span> <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                                            <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">minor</span><span class="p">)</span>
                                        <span class="k">else</span><span class="p">:</span>
                                            <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span><span class="p">)</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="c1"># Only mixed left</span>
                            <span class="k">for</span> <span class="n">g</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">minor</span> <span class="ow">in</span> <span class="n">indels</span><span class="p">:</span>
                                <span class="k">if</span> <span class="n">g</span> <span class="o">==</span> <span class="n">gene</span> <span class="ow">and</span> <span class="p">(</span>
                                    <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                                    <span class="ow">or</span> <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                                <span class="p">):</span>
                                    <span class="k">if</span> <span class="s2">&quot;mixed&quot;</span> <span class="ow">in</span> <span class="n">m</span><span class="p">:</span>
                                        <span class="n">matched</span> <span class="o">=</span> <span class="kc">True</span>
                                        <span class="k">if</span> <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                                            <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">minor</span><span class="p">)</span>
                                        <span class="k">else</span><span class="p">:</span>
                                            <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span><span class="p">)</span>
                        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">matched</span>

                <span class="k">elif</span> <span class="s2">&quot;?&quot;</span> <span class="ow">in</span> <span class="n">mut</span><span class="p">:</span>
                    <span class="c1"># Specific wildcard SNP, so match on everything except the alt</span>
                    <span class="n">mut</span> <span class="o">=</span> <span class="n">mut</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
                    <span class="k">for</span> <span class="n">g</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">minor</span> <span class="ow">in</span> <span class="n">snps</span><span class="p">:</span>
                        <span class="k">if</span> <span class="p">(</span>
                            <span class="n">g</span> <span class="o">==</span> <span class="n">gene</span>
                            <span class="ow">and</span> <span class="n">m</span><span class="p">[:</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">mut</span>
                            <span class="ow">and</span> <span class="p">(</span>
                                <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                                <span class="ow">or</span> <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                            <span class="p">)</span>
                        <span class="p">):</span>
                            <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="kc">True</span>
                            <span class="k">if</span> <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                                <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">minor</span><span class="p">)</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span><span class="p">)</span>

                <span class="k">elif</span> <span class="n">large_del_re</span><span class="o">.</span><span class="n">fullmatch</span><span class="p">(</span><span class="n">mutation</span><span class="p">):</span>
                    <span class="k">for</span> <span class="n">g</span><span class="p">,</span> <span class="n">m</span><span class="p">,</span> <span class="n">minor</span> <span class="ow">in</span> <span class="n">large_dels</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">g</span> <span class="o">==</span> <span class="n">gene</span> <span class="ow">and</span> <span class="p">(</span>
                            <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                            <span class="ow">or</span> <span class="p">(</span><span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">rule_is_minor</span><span class="p">)</span>
                        <span class="p">):</span>
                            <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="kc">True</span>
                            <span class="k">if</span> <span class="n">minor</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                                <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">minor</span><span class="p">)</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">g</span> <span class="o">+</span> <span class="s2">&quot;@&quot;</span> <span class="o">+</span> <span class="n">m</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Exact part, so check for it&#39;s existance</span>
                    <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">joined</span>
                    <span class="k">if</span> <span class="n">check</span><span class="p">:</span>
                        <span class="c1"># Exists, so add it to this multi</span>
                        <span class="n">this_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mutation</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">check</span><span class="p">:</span>
                    <span class="n">multi_match</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">this_match</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="c1"># Give up at first hurdle</span>
                    <span class="k">break</span>
            <span class="k">if</span> <span class="n">check</span><span class="p">:</span>
                <span class="c1"># Add all if valid</span>
                <span class="n">partials</span> <span class="o">=</span> <span class="n">multi_match</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">multi_match</span><span class="p">[</span><span class="mi">1</span><span class="p">:]:</span>
                    <span class="n">these_partials</span> <span class="o">=</span> <span class="p">[]</span>
                    <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">item</span><span class="p">:</span>
                        <span class="n">this_partial</span> <span class="o">=</span> <span class="p">[]</span>
                        <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">partials</span><span class="p">:</span>
                            <span class="n">this_partial</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">p</span> <span class="o">+</span> <span class="s2">&quot;&amp;&quot;</span> <span class="o">+</span> <span class="n">mutation</span><span class="p">)</span>
                        <span class="n">these_partials</span> <span class="o">+=</span> <span class="n">this_partial</span>
                    <span class="n">partials</span> <span class="o">=</span> <span class="n">these_partials</span>

                <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">partials</span><span class="p">:</span>
                    <span class="n">new_mutations</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="kc">None</span><span class="p">,</span> <span class="n">p</span><span class="p">))</span>

        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Exact multi, so just check for existance</span>
            <span class="k">for</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">multi</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&amp;&quot;</span><span class="p">):</span>
                <span class="n">check</span> <span class="o">=</span> <span class="n">check</span> <span class="ow">and</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">joined</span>
            <span class="k">if</span> <span class="n">check</span><span class="p">:</span>
                <span class="c1"># This exact multi mutation exists, so add it to the mutations list</span>
                <span class="n">new_mutations</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="kc">None</span><span class="p">,</span> <span class="n">multi</span><span class="p">))</span>
    <span class="k">if</span> <span class="n">just_joined</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">new_mutations</span>
    <span class="k">return</span> <span class="n">mutations_</span> <span class="o">+</span> <span class="n">new_mutations</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h2 id="gnomonicus_lib.write_mutations_csv" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">write_mutations_csv</span><span class="p">(</span><span class="n">mutations</span><span class="p">,</span> <span class="n">path</span><span class="p">,</span> <span class="nb">filter</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></code>

</h2>


    <div class="doc doc-contents ">

        <p>Prep and write the mutations CSV to the given filepath.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>mutations</code>
            </td>
            <td>
                  <code><span title="pandas.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Muations CSV</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>path</code>
            </td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Path to write to</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>filter</code>
            </td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Whether to filter nucleotide changes</p>
              </div>
            </td>
            <td>
                  <code>True</code>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>gnomonicus/gnomonicus_lib.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">write_mutations_csv</span><span class="p">(</span>
    <span class="n">mutations</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">path</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="nb">filter</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">True</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Prep and write the mutations CSV to the given filepath.</span>

<span class="sd">    Args:</span>
<span class="sd">        mutations (pd.DataFrame): Muations CSV</span>
<span class="sd">        path (str): Path to write to</span>
<span class="sd">        filter (bool, optional): Whether to filter nucleotide changes</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Reorder the columns</span>
    <span class="n">mutations</span> <span class="o">=</span> <span class="n">mutations</span><span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;uniqueid&quot;</span><span class="p">,</span>
            <span class="s2">&quot;gene&quot;</span><span class="p">,</span>
            <span class="s2">&quot;mutation&quot;</span><span class="p">,</span>
            <span class="s2">&quot;ref&quot;</span><span class="p">,</span>
            <span class="s2">&quot;alt&quot;</span><span class="p">,</span>
            <span class="s2">&quot;nucleotide_number&quot;</span><span class="p">,</span>
            <span class="s2">&quot;nucleotide_index&quot;</span><span class="p">,</span>
            <span class="s2">&quot;gene_position&quot;</span><span class="p">,</span>
            <span class="s2">&quot;codes_protein&quot;</span><span class="p">,</span>
            <span class="s2">&quot;indel_length&quot;</span><span class="p">,</span>
            <span class="s2">&quot;indel_nucleotides&quot;</span><span class="p">,</span>
            <span class="s2">&quot;amino_acid_number&quot;</span><span class="p">,</span>
            <span class="s2">&quot;amino_acid_sequence&quot;</span><span class="p">,</span>
            <span class="s2">&quot;number_nucleotide_changes&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span>

    <span class="c1"># As we have concated several dataframes, the index is 0,1,2,0,1...</span>
    <span class="c1"># Reset it so that we can use it to delete</span>
    <span class="n">mutations</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">mutations_</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">mutations</span><span class="p">)</span>
    <span class="k">if</span> <span class="nb">filter</span><span class="p">:</span>
        <span class="c1"># Filter out nucleotide variants from synonymous mutations to avoid duplication of data</span>
        <span class="n">to_drop</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">idx2</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">mutations_</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="n">row</span><span class="p">[</span><span class="s2">&quot;codes_protein&quot;</span><span class="p">]</span>
                <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
                <span class="ow">and</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;alt&quot;</span><span class="p">]</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span>
            <span class="p">):</span>
                <span class="c1"># Protein coding so check if nucleotide within coding region</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;ref&quot;</span><span class="p">])</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                    <span class="c1"># Nucleotide SNP</span>
                    <span class="n">to_drop</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">idx2</span><span class="p">)</span>
        <span class="n">mutations_</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">to_drop</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># Save it as CSV</span>
    <span class="n">mutations_</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">path</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
      
      <nav class="md-footer__inner md-grid" aria-label="Footer" >
        
          
          <a href="../" class="md-footer__link md-footer__link--prev" aria-label="Previous: gnomonicus">
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
            </div>
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Previous
              </span>
              <div class="md-ellipsis">
                gnomonicus
              </div>
            </div>
          </a>
        
        
          
          <a href="../../bin/" class="md-footer__link md-footer__link--next" aria-label="Next: CLI">
            <div class="md-footer__title">
              <span class="md-footer__direction">
                Next
              </span>
              <div class="md-ellipsis">
                CLI
              </div>
            </div>
            <div class="md-footer__button md-icon">
              
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11z"/></svg>
            </div>
          </a>
        
      </nav>
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.footer", "navigation.instant", "navigation.sections", "navigation.tabs", "navigation.tracking"], "search": "../../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "jeremy"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.f1b6f286.min.js"></script>
      
    
  </body>
</html>